{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing the modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [],
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/python\n",
    "import sys\n",
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "import copy\n",
    "import operator\n",
    "import glob\n",
    "#import pprint\n",
    "import argparse\n",
    "import datetime\n",
    "import queue\n",
    "import multiprocessing\n",
    "from alive_progress import alive_bar\n",
    "import numpy.matlib\n",
    "import numpy as np     # For Mathmatical (Algebra) Operation\n",
    "from sys import platform   # To verify the Operating Sys Win or Linux\n",
    "from Bio import SeqIO  # For Bio Python Sequence Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Blastp = /usr/bin/blastp\n",
    "Blastp = \"blastp\"\n",
    "Blastp_data =  './blastp_data/'\n",
    "cpu_count = 1 # Default value\n",
    "blastp_matrix = \"BLOSUM62\" \n",
    "genome = \"\"\n",
    "infinite_loop = 70\n",
    "inflation_factor = 1.4\n",
    "#mode = None\n",
    "mode = '1' #Change Later\n",
    "Cluster_out = \"./cluster_out/\"\n",
    "#threshold_score : 0\n",
    "threshold_score =0\n",
    "Save_raw_blastp_score = False\n",
    "Score_file = \"./score_file\"\n",
    "Species = \"./species/\"\n",
    "species = \"./species/\"\n",
    "verbose = False\n",
    "Log_file_name = \"log_files/\"\n",
    "save_raw_blastp_score = \"score_file/\"\n",
    "\n",
    "mode = '1'\n",
    "matrix = \"BLOSUM62\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. Calling the Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 1. matrix_name()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": [],
    "lang": "en"
   },
   "outputs": [],
   "source": [
    "def matrix_name():\n",
    "    \"\"\"This Function will Return the Matrix name choosed by User.\n",
    "    BLOSUM45 ,  BLOSUM62 , BLOSUM82 \"\"\"\n",
    "    print(\"BLOcks SUbstitution Matrix (BLOSUM) is a Substitution matrix used for sequence alignment of Proteins\")\n",
    "    print(\"\"\"\\n1. BLOSUM45 :-For more distantly related Proteins alignment DataBase\\n2. BLOSUM62 :- MidRange Seq with more than 62%similarity\\\n",
    "         \\n3. BLOSUM82 :- More related Proteins\\nOther Keys will exit the Program\"\"\")\n",
    "    metrix_num = input(\"\\nEnter a matrix number: \")\n",
    "    if metrix_num not in (\"1\", \"2\", '3'):\n",
    "        print(\"Wrong input *%s*Sorry not in list\\n\" %\n",
    "              metrix_num, \"*\"*20, \"Good Bye\", \"*\"*20, \"\\n\")\n",
    "        sys.exit(1)\n",
    "    if metrix_num == \"1\":\n",
    "        return \"BLOSUM45\"\n",
    "    if metrix_num == \"2\":\n",
    "        return \"BLOSUM62\"  # 62 is not availiable currently\n",
    "    if metrix_num == \"3\":\n",
    "        return \"BLOSUM82\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 2. Query_Sequence(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def query_seq_(genome):\n",
    "    \"\"\"This Function Read Fasta (Genome) file and return as a list Format with gene Position and Sequence Developed by Krish\"\"\"\n",
    "    try:\n",
    "        return [(seq_record.id+\"\\n\"+str(seq_record.seq)) for seq_record in SeqIO.parse(genome, \"fasta\")]\n",
    "        # To understand this Function Bio Python library needs to be studied\n",
    "    except IOError as err:\n",
    "        print(str(err))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": [],
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "def query_sequence(genome):\n",
    "    \"\"\"This Function Read Fastaq Files and return as a list Format with gene Position as a index\n",
    "    This Function create a list of Seprate Query Sequence\"\"\"\n",
    "    # This Function is created by using Bio Python we will test later\n",
    "    gene_seq = \"\"\n",
    "    gene_seq_list = []\n",
    "    try:\n",
    "        with open(Species+genome) as gene:\n",
    "            for each_line in gene:\n",
    "                if \">\" in each_line:\n",
    "                    if gene_seq != \"\":\n",
    "                        gene_seq_list.append(gene_seq)\n",
    "                        gene_seq = \"\"\n",
    "                gene_seq = gene_seq+each_line\n",
    "            gene_seq_list.append(gene_seq)\n",
    "            #print(\"query_sequence() Run Successfully\")\n",
    "            return gene_seq_list\n",
    "\n",
    "    except IOError as err:\n",
    "        print(\"IOError occurred in query_sequence function : \" + str(err))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!ls ./species/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = \"chicken.faa\"\n",
    "gene_seq = []\n",
    "gene_seq_list = []\n",
    "try :\n",
    "    with open(species+genome)as gene:\n",
    "        for each_line in gene:\n",
    "            if \">\" in each_line:\n",
    "                if gene_seq !=\"\":\n",
    "                    gene_seq_list.append(gene_seq)\n",
    "                    #print(gene_seq)\n",
    "                    gene_seq=\"\"\n",
    "            gene_seq = gene_seq+each_line\n",
    "        gene_seq_list.append(gene_seq)\n",
    "except IOError as err:\n",
    "    print(\"File Not Found\")\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene_seq_list[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "###### 3.write_query(query, parallel_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "code_folding": [],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def write_query(query, parallel_num):\n",
    "    \"This Function Write Query with file Name query+ parallel_num in same directory and raise IO error if Error rises\"\n",
    "    # print(\"write_query(query,parallel_num)\")\n",
    "    try:\n",
    "        with open(\"./query/query_\"+str(parallel_num), \"w\") as write_query:\n",
    "            write_query.write(query)\n",
    "    except IOError as err:\n",
    "        print(\"IOError occurred in write_query function : \" + str(err))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 4. Run_Blast( Subject , parallel_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def run_blast(subject, parallel_num):\n",
    "    \"\"\"\n",
    "    By this Function it will create a Pipe line to run Blastp in Computer by input Parameter.\n",
    "    Subject is whole Genome and parallel_number is query file Created by early step.\n",
    "    Running a big file is time Consuming.The parameters passed in blastp are\n",
    "    The Output Format 10 --> Comma Separated Values\n",
    "        >> qseqid --> Query Seq ID\n",
    "        >> ssequid --> Subject of Seq. id ID\n",
    "        >> Score -->Raw Score\n",
    "        >> length-->Alignment length\n",
    "    # query subject is inside query Folder\n",
    "    # run_blastp =subprocess.Popen(cmd,stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    # run_blastp = subprocess.run(cmd,shell=True ,capture_output = True,text = True)\n",
    "    # output Format 10 qseqid query (e.g. gene sequence id , sseqid subject (e.g. reference genome) genome id)\"\"\"\n",
    "\n",
    "    subject = Species+subject  # ** later remove this\n",
    "    cmd = [\"blastp\", \"-query\", \"./query/query\"+str(parallel_num), \"-subject\", subject,\n",
    "           \"-matrix\", blastp_matrix, \"-outfmt\", \"10 qseqid sseqid score length\"]\n",
    "\n",
    "    run_blastp = subprocess.run(cmd, capture_output=True, text=True)\n",
    "    # If you want to run in Windows pass shell = True\n",
    "    if run_blastp.stderr:\n",
    "        print(\"Check Blastp Function 230\")\n",
    "        print(run_blastp.stderr)\n",
    "        with open (\"error.txt\",\"a+\")as log:\n",
    "            log.write(run_blastp.stderr)\n",
    "        sys.exit(2)\n",
    "    return run_blastp.stdout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Check Blastp Function 230\n",
      "Command line argument error: Argument \"query\". File is not accessible:  `./query/query2'\n",
      "\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[1;31mSystemExit\u001b[0m\u001b[1;31m:\u001b[0m 2\n"
     ]
    }
   ],
   "source": [
    "display(run_blast(\"AAE\",2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def Run_Blast1(subject, parallel_num):\n",
    "    \"By this Function it will create a Pipe line to run Blastp in Computer by input Parameter Subject is whole Genome and parallel_number is query file Created by early step\"\n",
    "    #print(\"Runing Run_Blast(subject,parallel_number)\")\n",
    "    # print(\"Passed Parallel Number is \",parallel_num) #parallel_number is the cpu_count\n",
    "    subject = Species+subject\n",
    "    cmd = [\"blastp\", \"-query\", \"./query/query_\"+str(parallel_num), \"-subject\", subject,\n",
    "           \"-matrix\", blastp_matrix, \"-outfmt\", \"10 qseqid sseqid score length\"]\n",
    "    run_blastp = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    \n",
    "    return run_blastp.communicate()[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 5. same_species_forward_best_hit(blastp_score):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "code_folding": [
     0
    ],
    "lang": "en"
   },
   "outputs": [],
   "source": [
    "def same_species_forward_best_hit(blastp_score):\n",
    "    \"\"\"Search the forward best hit among the blastp scores of same species.\n",
    "    Because there are an duplicated genes in a same genome.\n",
    "    blstp_score file is return value of run_blast() Function\n",
    "    When the blastp score compare with blastp score of duplicate gene, if score and length are same, blasp score of duplicated gene is added to a second best score.\"\"\"\n",
    "    #print(blastp_score)\n",
    "    blastp_score_split_list = []\n",
    "    temp_best_score = ['-1', '-1', '-1']\n",
    "    second_temp_best_score = []\n",
    "    best_score = []\n",
    "    second_best_score = []\n",
    "    blastp_score_split = blastp_score.split(\"\\n\")\n",
    "    # delete of [''] in the last index  ex) ['gi,gi,1,1','gi,gi,2,2','']\n",
    "    del blastp_score_split[-1]\n",
    "    for i in blastp_score_split:\n",
    "        blastp_score_element = i.split(',')\n",
    "        blastp_score_split_list.append(blastp_score_element)\n",
    "    # ex) k is ['gi|15605613|ref|NP_212986.1|', 'gi|15605613|ref|NP_212986.1|', '3702', '699']\n",
    "    for k in blastp_score_split_list:\n",
    "        if k[0] == k[1]:  # if the Position (gene) is Same\n",
    "            best_score.append(k)\n",
    "        elif k[0] != k[1]:\n",
    "            if int(k[2]) > int(temp_best_score[2]):  # Compare score\n",
    "                temp_best_score = k\n",
    "            elif int(k[2]) == int(temp_best_score[2]):  # Is Equal\n",
    "                if int(k[3]) > int(temp_best_score[3]):  # compare length\n",
    "                    temp_best_score = k\n",
    "                elif int(k[3]) == int(temp_best_score[3]):\n",
    "                    second_temp_best_score.append(k)\n",
    " #    print (\"############ temp best score ############\", temp_best_score)\n",
    "    second_best_score.append(temp_best_score)\n",
    "    for j in second_temp_best_score:\n",
    "        if j[2] == temp_best_score[2] and j[3] == temp_best_score[3]:\n",
    "            second_best_score.append(j)\n",
    "    for m in second_best_score:\n",
    "        if (best_score[0][2] == m[2] and int(best_score[0][3]) <= int(m[3])) or int(best_score[0][2]) < int(m[2]):\n",
    "             # '104' < '23' is True because of string. So the int function is used.\n",
    "            best_score.append(m)\n",
    "    #print(\"best Score is\", best_score)\n",
    "    return best_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Same_Species_Forward_Best_HIt(a)\n",
    "#a = str(blastp_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 6. forward_best_hit(blastp_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def forward_best_hit(blastp_score):\n",
    "    \"\"\"Search the forward best hit among the blastp scores of same species.\"\"\"\n",
    "    #print(\"Rnunning GetForward BestHit\")\n",
    "    blastp_score_split_list = []\n",
    "    temp_best_score = ['-1', '-1', '-1']\n",
    "    second_temp_best_score = []\n",
    "    best_score = []\n",
    "    blastp_score_split = blastp_score.split(\"\\n\")\n",
    "    # delete of ['']   ex) ['gi,gi,1,1','gi,gi,2,2','']\n",
    "    del blastp_score_split[-1]\n",
    "    for i in blastp_score_split:\n",
    "        blastp_score_element = i.split(',')\n",
    "        blastp_score_split_list.append(blastp_score_element)\n",
    "\n",
    "    for k in blastp_score_split_list:\n",
    "        # ex) k is ['gi|15605613|ref|NP_212986.1|', 'gi|15605613|ref|NP_212986.1|', '3702', '699']\n",
    "        # print \">>>>>>>>>>>>>>>forward_best_hit   k\", k\n",
    "        if int(k[2]) > int(temp_best_score[2]):  # Compare the Score of the Blast\n",
    "            temp_best_score = k\n",
    "        elif int(k[2]) == int(temp_best_score[2]):\n",
    "            if int(k[3]) > int(temp_best_score[3]):  # Compare the Lenth of Sequence\n",
    "                temp_best_score = k\n",
    "            elif int(k[3]) == int(temp_best_score[3]):\n",
    "                second_temp_best_score.append(k)\n",
    "  #print \"############ temp best score ############\", temp_best_score\n",
    "    best_score.append(temp_best_score)\n",
    "    for j in second_temp_best_score:\n",
    "        if j[2] == temp_best_score[2] and j[3] == temp_best_score[3]:\n",
    "            best_score.append(j)\n",
    "    return best_score, blastp_score_split_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#a,b  = Forward_Best_HIt(blastp_score.decode())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 7. division_parallel_query(queryV,query_division_value, cpu_count , queryV_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def division_parallel_query(query_v, query_division_value, cpu_count, query_v_len):\n",
    "    \"query_v is list of queries, query_division_value is , cpu_count is the Number of CPU User Entered and query_v_len is the length of the query\"\n",
    "    parallel_query = []\n",
    "    parallel_query_start = 0\n",
    "    if query_v_len % cpu_count == 0:  # perfect division\n",
    "        for i in range(cpu_count):\n",
    "            i += 1\n",
    "            if parallel_query_start == 0:\n",
    "                parallel_query.append(\n",
    "                    query_v[int(parallel_query_start):i*int(query_division_value)])\n",
    "                parallel_query_start += 1\n",
    "            else:\n",
    "                parallel_query.append(query_v[int(\n",
    "                    parallel_query_start)*int(query_division_value):i*int(query_division_value)])\n",
    "                parallel_query_start += 1\n",
    "    else:  # imperfect division\n",
    "        for i in range(cpu_count):\n",
    "            i += 1\n",
    "            if parallel_query_start == 0:\n",
    "                parallel_query.append(\n",
    "                    query_v[int(parallel_query_start):i*int(query_division_value)])\n",
    "                parallel_query_start += 1\n",
    "            elif i < cpu_count:\n",
    "                parallel_query.append(query_v[int(\n",
    "                    parallel_query_start)*int(query_division_value):i*int(query_division_value)])\n",
    "                parallel_query_start += 1\n",
    "            elif i == cpu_count:\n",
    "                parallel_query.append(\n",
    "                    query_v[int(parallel_query_start)*int(query_division_value):])\n",
    "                parallel_query_start += 1\n",
    "\n",
    "    return parallel_query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 8. Run_Parallel_Query(species_of_query, species_of_subject, queryV, parallel_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def run_parallel_query(species_of_query, species_of_subject, query_v, parallel_num):\n",
    "    \"\"\" run_parallel_query(i ,k , query_v , cpu_count) i and k are user selected number in a list Format\n",
    "    Run the following functions. write_query, run_blast, same_species_forward_best_hit, forward_best_hit\n",
    "    Save the files which are oneway_threshold_best_hit, second_oneway_threshold_best_hit, \n",
    "    blastp_score_split_list and raw_blastp_score (optional) by each species. \n",
    "    parallel_num is the Number of CPU selected by the user if CPu 1 selected then 1 \"\"\"\n",
    "\n",
    "    #print(\"run_parallel_query Running\")\n",
    "    global selected_number, selected_species_dic\n",
    "    # bar = Bar('Processing '+str(parallel_num), max = len(query_v)) #progressing bar setting , Creating a Object\n",
    "    # bar is not Supported in Python 3    We use alive_bar instead\n",
    "    with alive_bar(len(query_v)) as bar:  # declare your set of items for loop\n",
    "        for j in query_v:\n",
    "            # bar.next() #progressing bar not supported\n",
    "            bar()# Call after Consuming One Item\n",
    "            write_query(j, parallel_num)  # if 1 Added Here Add also to Run\n",
    "        # This Function Only Write a file with j name and parallel_num i.e CPU Count\n",
    "            blastp_score = run_blast(\n",
    "                selected_species_dic[species_of_subject], parallel_num)\n",
    "        # \" Return Byte File type, and this is only for one gene position query and Whole Genome of species\"\n",
    "            if blastp_score != '':  # if blastp run successfully\n",
    "                best_score, blastp_score_split_list = forward_best_hit(\n",
    "                    blastp_score.decode())  # .decode() Convert in to str format\n",
    "                # ex) AAE == AAE. It will save best_score without reversing run_blast.\n",
    "                if species_of_query == species_of_subject:\n",
    "                    same_species_forward_best_score = same_species_forward_best_hit(\n",
    "                        blastp_score.decode())\n",
    "                    for best_score_element in same_species_forward_best_score:\n",
    "                        # ex) [A1 of AAE, A1 of AAE, 30]\n",
    "                        if best_score_element[0] == best_score_element[1]:\n",
    "                            with open(Score_file+selected_species_dic[species_of_query] + \"_\" +\n",
    "                                      selected_species_dic[species_of_subject]\n",
    "                                      + \"_oneway_threshold_best_hit_Score\"\n",
    "                                      + str(threshold_score), \"a\") as oneway_threshold_best_hit:\n",
    "                                save_best_score = selected_species_dic[species_of_query]+\"_\"+best_score_element[0].split(\"\\s\")[0]\\\n",
    "                                    + \" \"+selected_species_dic[species_of_subject]+\"_\"+best_score_element[1].split(\n",
    "                                        \"\\s\")[0]+\" \"+best_score_element[2]+\"\\n\"\n",
    "                                # best_score_element[0].split(\"|\") ==> ['gi', '15642790', 'ref', 'NP_227831.1', '']\n",
    "                                oneway_threshold_best_hit.write(\n",
    "                                    save_best_score)\n",
    "                        else:  # ex) [A1 of AAE, A2 of AAE, 30]\n",
    "                            with open(Score_file+selected_species_dic[species_of_query]+\"_\"\n",
    "                                      + selected_species_dic[species_of_subject]\n",
    "                                      + \"_second_oneway_threshold_best_hit_Score\"\n",
    "                                      + str(threshold_score), \"a\") as second_oneway_threshold_best_hit:\n",
    "                                second_save_best_score = selected_species_dic[species_of_query]\\\n",
    "                                    + \"_\"+best_score_element[0].split(\"\\s\")[0]\\\n",
    "                                    + \" \"+selected_species_dic[species_of_subject]+\"_\"\\\n",
    "                                    + best_score_element[1].split(\"\\s\")[0]+\" \"+best_score_element[2]+\"\\n\"\n",
    "                                second_oneway_threshold_best_hit.write(\n",
    "                                    second_save_best_score)\n",
    "                else:  # If species_of_query not equal with species_of_subject, run reversing run_blast\n",
    "                    for best_score_element in best_score:\n",
    "                        if not '-1' in best_score_element:\n",
    "                            with open(Score_file+selected_species_dic[species_of_query]\n",
    "                                      + \"_\" +\n",
    "                                      selected_species_dic[species_of_subject] +\n",
    "                                      \"_\"+\"best_score_S\"\n",
    "                                      + str(threshold_score)+\"_\"+str(parallel_num), \"a\") as save_best_hit:\n",
    "                                best_score_save = selected_species_dic[species_of_query]+\"_\"\\\n",
    "                                    + best_score_element[0].split(\"\\s\")[0]+\" \"+selected_species_dic[species_of_subject]\\\n",
    "                                    + \"_\"+best_score_element[1].split(\n",
    "                                        \"\\s\")[0]+\" \"+best_score_element[2]+\" \"+best_score_element[3]+\"\\n\"\n",
    "                                save_best_hit.write(best_score_save)\n",
    "\n",
    "                    for blastp_score_split_list_element in blastp_score_split_list:\n",
    "                        with open(Score_file+selected_species_dic[species_of_query]+\"_\"\n",
    "                                  + selected_species_dic[species_of_subject]+\"_\"+\"blastp_score_split_list_S\"\n",
    "                                  + str(threshold_score)+\"_\"+str(parallel_num), \"a\") as save_blastp_score_split_list:\n",
    "                            blastp_score_split_list_save = selected_species_dic[species_of_query]+\"_\"\\\n",
    "                                + blastp_score_split_list_element[0].split(\"\\s\")[0]\\\n",
    "                                + \" \"+selected_species_dic[species_of_subject]+\"_\"+blastp_score_split_list_element[1].split(\"\\s\")[0]\\\n",
    "                                + \" \" + \\\n",
    "                                blastp_score_split_list_element[2]+\" \" \\\n",
    "                                + blastp_score_split_list_element[3]+\"\\n\"\n",
    "                            save_blastp_score_split_list.write(\n",
    "                                blastp_score_split_list_save)\n",
    "\n",
    "            else:\n",
    "                print(\"No value in blastp_score\")\n",
    "            if save_raw_blastp_score:\n",
    "                with open(Score_file+selected_species_dic[species_of_query]\n",
    "                          + \"_\"+selected_species_dic[species_of_subject]+\"_S\"\n",
    "                          + str(threshold_score)+\"_\"+str(parallel_num), \"a\") as save_blastp:\n",
    "                    save_blastp.write(blastp_score)\n",
    "    # bar.finish() # progressing bar finish\n",
    "    return  # None\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 9.oneway_threshold_best_hit(mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "code_folding": [
     0
    ],
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "def oneway_threshold_best_hit(mode):\n",
    "    \"\"\" This Function accept the mode and Run program and return backward_best_hit_work_list mode 1 and 2 is supported \"\"\"\n",
    "    #print(\"oneway_threshold_best_hit running\")\n",
    "    global user_selected_number, cpu_count\n",
    "    b_info = \"Running the blastp & forward best hit searches \"\n",
    "    process_list = []\n",
    "    backward_best_hit_work_list = []\n",
    "    if \"1\" in mode:\n",
    "        \"\"\"We have 3 Mode 1 is for blastp, Mode 2 is for BLASTP using precalcualted data and Mode 3 is for clustering\"\"\"\n",
    "        for i in user_selected_number:  # Select species to write query \n",
    "            query_v = query_sequence(selected_species_dic[i])\n",
    "            # query_v is a list Format  with a position gene id , Seq\n",
    "            # User Selected  [1, 3, 5] is list of User input\n",
    "            for k in user_selected_number:\n",
    "                if k < i:  # gene ====> query 1->1 1->2 1->3 2->2 2->3  Forward checking will skip the same or less\n",
    "                    continue\n",
    "                else:\n",
    "                    print(b_info+\" between %s genome and %s genome\"\n",
    "                          % (selected_species_dic[i], selected_species_dic[k]))\n",
    "                    # this will Create a loop though 2 times\n",
    "                    # length of Genome gene ID in Sequence File\n",
    "                    query_v_len = len(query_v)\n",
    "                    if cpu_count == 1:\n",
    "                        #\"No Parallel Computing while cpu count == 1\"\n",
    "                        blastp_time_start = time.time()\n",
    "                        run_parallel_query(i, k, query_v, cpu_count)\n",
    "                        #\"i is first species k is second species number queryv is list file of i Position genome\"\n",
    "                        blastp_time_end = time.time()\n",
    "                        print(b_info+ \"took %.2f minutes\" % (\n",
    "                            (blastp_time_end-blastp_time_start)/60))\n",
    "                    else:\n",
    "                        # If the number of query_v_len is less than cpu_count, Remark will select the number of query_v_len.\n",
    "                        if query_v_len < cpu_count:\n",
    "                            #\"because the cpu_count will seprate list items from query_v file.In general this will not happened because length of query is always long\"\n",
    "                            blastp_time_start = time.time()\n",
    "                            # 1 is query_division_value. Because query_v_len / query_v_len(=cpu_count) is 1.\n",
    "                            parallel_query = division_parallel_query(\n",
    "                                query_v, 1, query_v_len, query_v_len)\n",
    "                            for m in range(query_v_len):\n",
    "                                process = multiprocessing.Process(\n",
    "                                    target=run_parallel_query, args=(i, k, parallel_query[m], m+1))\n",
    "                                # args( i => species of query , k => species of subject, m+1 => cpu_count ex) 1, 2 ...)\n",
    "                                process_list.append(process)\n",
    "                                process.start()\n",
    "                            for n in process_list:\n",
    "                                n.join()\n",
    "                            blastp_time_end = time.time()\n",
    "                            print(b_info+\"took %.2f minutes \" % (\n",
    "                                (blastp_time_end-blastp_time_start)/60))\n",
    "                        else:\n",
    "                            #\"this wiil run if cpu_count is more than 1\"\n",
    "                            blastp_time_start = time.time()\n",
    "                            query_division_value = query_v_len / cpu_count\n",
    "                            parallel_query = division_parallel_query(\n",
    "                                query_v, query_division_value, cpu_count, query_v_len)\n",
    "                            for m in range(cpu_count):\n",
    "                                process = multiprocessing.Process(\n",
    "                                    target=run_parallel_query, args=(i, k, parallel_query[m], m+1))\n",
    "                                # args( i => species of query , k => species of subject, m+1 => cpu_count ex) 1, 2 ...)\n",
    "                                process_list.append(process)\n",
    "                                process.start()\n",
    "                            for n in process_list:\n",
    "                                n.join()\n",
    "                            blastp_time_end = time.time()\n",
    "                            print(b_info+\"took %.2f minutes \" % (\n",
    "                                (blastp_time_end-blastp_time_start)/60))\n",
    "                if not i == k:\n",
    "                    backward_best_hit_work_list.append((i, k, query_v_len))\n",
    "                    print(\"Backward\")\n",
    "    elif \"2\" in mode:\n",
    "        \"\"\"If blast Files exist then this will pass the blast time other wise blast will run.\n",
    "        This will reduce the time for blast and increase the speed of the system.\n",
    "        \"\"\"\n",
    "        for i in user_selected_number:  # Select species to write query\n",
    "            query_v = query_sequence(selected_species_dic[i])\n",
    "            for k in user_selected_number:  # Select of subject\n",
    "                if Blastp_data+selected_species_dic[i]+\"_\"+selected_species_dic[k]\\\n",
    "                    + \"_oneway_threshold_best_hit_Score\"\\\n",
    "                        + str(threshold_score) in precalculated_data_list:\n",
    "                    used_precalculated_data_list.append(\n",
    "                        selected_species_dic[i]+\"_\"+selected_species_dic[k])\n",
    "                    #used_precalcualted _data_list is created before calling this function\n",
    "                    #if files exist then passed\n",
    "                    continue #File will skkipped\n",
    "                else:\n",
    "                    if k < i:  # gene ====> query 1->1 1->2 1->3 2->2 2->3\n",
    "                        continue\n",
    "                    else:\n",
    "                        \"same as Mode 1, if Precalculated file doesnot exist\"\n",
    "                        print(b_info+ \" between %s genome and %s genome\" % (\n",
    "                            selected_species_dic[i], selected_species_dic[k]))\n",
    "                        query_v_len = len(query_v)\n",
    "                        if cpu_count == 1:\n",
    "                            blastp_time_start = time.time()\n",
    "                            run_parallel_query(i, k, query_v, cpu_count)\n",
    "                            blastp_time_end = time.time()\n",
    "                            print(b_info+ \"took %.2f minutes\" % (\n",
    "                                (blastp_time_end-blastp_time_start)/60))\n",
    "                        else:\n",
    "                            # If the number of query_v_len is less than cpu_count, Remark will select the number of query_v_len.\n",
    "                            if query_v_len < cpu_count:\n",
    "                                blastp_time_start = time.time()\n",
    "                                # 1 is query_division_value. Because query_v_len / query_v_len(=cpu_count) is 1.\n",
    "                                parallel_query = division_parallel_query(\n",
    "                                    query_v, 1, query_v_len, query_v_len)\n",
    "                                for m in range(query_v_len):\n",
    "                                    process = multiprocessing.Process(\n",
    "                                        target=run_parallel_query,\n",
    "                                        args=(i, k, parallel_query[m], m+1))\n",
    "                                    # args( i => species of query , k => species of subject, m+1 => cpu_count ex) 1, 2 ...)\n",
    "                                    process_list.append(process)\n",
    "                                    process.start()\n",
    "                                for n in process_list:\n",
    "                                    n.join()\n",
    "                                blastp_time_end = time.time()\n",
    "                                print(b_info+\" took %.2f minutes\" % (\n",
    "                                    (blastp_time_end-blastp_time_start)/60))\n",
    "                            else:\n",
    "                                blastp_time_start = time.time()\n",
    "                                query_division_value = query_v_len / cpu_count\n",
    "                                parallel_query = division_parallel_query(\n",
    "                                    query_v, query_division_value, cpu_count, query_v_len)\n",
    "                                for m in range(cpu_count):\n",
    "                                    process = multiprocessing.Process(\n",
    "                                        target=run_parallel_query,\n",
    "                                        args=(i, k, parallel_query[m], m+1))\n",
    "                                    # args( i => species of query , k => species of subject, m+1 => cpu_count ex) 1, 2 ...)\n",
    "                                    process_list.append(process)\n",
    "                                    process.start()\n",
    "                                for n in process_list:\n",
    "                                    n.join()\n",
    "                                blastp_time_end = time.time()\n",
    "                                print(b_info+ \" took %.2f minutes\" % (\n",
    "                                    (blastp_time_end-blastp_time_start)/60))\n",
    "                        new_calculated_data_list.append(\n",
    "                            selected_species_dic[i]+\"_\"+selected_species_dic[k])\n",
    "                        if not i == k:\n",
    "                            backward_best_hit_work_list.append(\n",
    "                                (i, k, query_v_len))\n",
    "    return backward_best_hit_work_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Running the blastp & forward best hit searches between AAE genome and AAE genome'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "UnicodeEncodeError",
     "evalue": "'cp949' codec can't encode character '\\u26a0' in position 1: illegal multibyte sequence",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\alive_progress\\progress.py\u001b[0m in \u001b[0;36malive_bar\u001b[1;34m(total, title, calibrate, **options)\u001b[0m\n\u001b[0;32m    264\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 265\u001b[1;33m         \u001b[1;32myield\u001b[0m \u001b[0mbar\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    266\u001b[0m     \u001b[1;32mfinally\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-12-11e03ed0f8d8>\u001b[0m in \u001b[0;36mrun_parallel_query\u001b[1;34m(species_of_query, species_of_subject, query_v, parallel_num)\u001b[0m\n\u001b[0;32m     21\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mblastp_score\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m:\u001b[0m  \u001b[1;31m# if blastp run successfully\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 22\u001b[1;33m                 best_score, blastp_score_split_list = forward_best_hit(\n\u001b[0m\u001b[0;32m     23\u001b[0m                     blastp_score.decode())  # .decode() Convert in to str format\n",
      "\u001b[1;31mNameError\u001b[0m: name 'forward_best_hit' is not defined",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mUnicodeEncodeError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-69-35decd896a0f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     12\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mcpu_count\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     13\u001b[0m                 \u001b[0mblastp_start_time\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 14\u001b[1;33m                 \u001b[0ma\u001b[0m\u001b[1;33m=\u001b[0m \u001b[0mrun_parallel_query\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mquery_v\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mcpu_count\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     15\u001b[0m                 \u001b[0mdisplay\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-12-11e03ed0f8d8>\u001b[0m in \u001b[0;36mrun_parallel_query\u001b[1;34m(species_of_query, species_of_subject, query_v, parallel_num)\u001b[0m\n\u001b[0;32m     83\u001b[0m                           \u001b[1;33m+\u001b[0m \u001b[1;34m\"_\"\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mselected_species_dic\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mspecies_of_subject\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m\"_S\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     84\u001b[0m                           + str(threshold_score)+\"_\"+str(parallel_num), \"a\") as save_blastp:\n\u001b[1;32m---> 85\u001b[1;33m                     \u001b[0msave_blastp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mblastp_score\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     86\u001b[0m     \u001b[1;31m# bar.finish() # progressing bar finish\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     87\u001b[0m     \u001b[1;32mreturn\u001b[0m  \u001b[1;31m# None\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\contextlib.py\u001b[0m in \u001b[0;36m__exit__\u001b[1;34m(self, type, value, traceback)\u001b[0m\n\u001b[0;32m    128\u001b[0m                 \u001b[0mvalue\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    129\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 130\u001b[1;33m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mthrow\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtraceback\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    131\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    132\u001b[0m                 \u001b[1;31m# Suppress StopIteration *unless* it's the same exception that\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\alive_progress\\progress.py\u001b[0m in \u001b[0;36malive_bar\u001b[1;34m(total, title, calibrate, **options)\u001b[0m\n\u001b[0;32m    273\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    274\u001b[0m         \u001b[0mend\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrun\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrun\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstats\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstats_end\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 275\u001b[1;33m         \u001b[0malive_repr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\alive_progress\\progress.py\u001b[0m in \u001b[0;36malive_repr\u001b[1;34m(spin)\u001b[0m\n\u001b[0;32m    116\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mline_len\u001b[0m \u001b[1;33m<\u001b[0m \u001b[0mrun\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlast_line_len\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    117\u001b[0m                 \u001b[0mclear_traces\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 118\u001b[1;33m             \u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__stdout__\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mline\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mspin\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;34m'\\r'\u001b[0m \u001b[1;32mor\u001b[0m \u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    119\u001b[0m             \u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__stdout__\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    120\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mUnicodeEncodeError\u001b[0m: 'cp949' codec can't encode character '\\u26a0' in position 1: illegal multibyte sequence"
     ]
    }
   ],
   "source": [
    "cpu_count = 1\n",
    "b_info = \"Running the blastp & forward best hit searches \"\n",
    "for i in user_selected_number:\n",
    "    query_v = query_sequence(selected_species_dic[i])\n",
    "    for k in user_selected_number:\n",
    "        if k < i:\n",
    "            continue\n",
    "        else:\n",
    "            display(b_info+ \"between %s genome and %s genome\"\\\n",
    "                   %(selected_species_dic[i],selected_species_dic[k]))\n",
    "            query_v_len = len(query_v)\n",
    "            if cpu_count == 1:\n",
    "                blastp_start_time = time.time()\n",
    "                a= run_parallel_query (i,k,query_v,cpu_count)\n",
    "                display(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 10. Backward_Best_Hit(args):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def backward_best_hit(args):\n",
    "    bck_info = \"Running the backward_best_hit\"\n",
    "    test =\" \"\n",
    "    print(bck_info)\n",
    "    species_of_query, species_of_subject, query_v_len = args\n",
    "    start_time_bbh = time.time()\n",
    "    forward_best_hit_score_list = []\n",
    "    blastp_score_split_list = []\n",
    "    print(bck_info+ \" between %s genome %s genome\" % (\n",
    "        selected_species_dic[species_of_query], selected_species_dic[species_of_subject]))\n",
    "    # If the number of query_v_len is less than cpu_count, the cpu_count is changed to query_v_len.\n",
    "    if query_v_len < cpu_count:\n",
    "        for parallel_num in range(query_v_len):\n",
    "            parallel_num += 1\n",
    "            with open(Score_file+selected_species_dic[species_of_query]+\"_\"\n",
    "                      + selected_species_dic[species_of_subject] + \"_\"+\"best_score_S\"\n",
    "                      + str(threshold_score)+\"_\"+str(parallel_num), \"r\") as best_hit_score:\n",
    "                for each_line in best_hit_score:\n",
    "                    split_each_line = each_line.split(\" \")\n",
    "                    split_each_line[2] = int(split_each_line[2])\n",
    "                    split_each_line[3] = int(split_each_line[3])\n",
    "                    forward_best_hit_score_list.append(split_each_line)\n",
    "            with open(Score_file+selected_species_dic[species_of_query]+\"_\"\n",
    "                      + selected_species_dic[species_of_subject] +\n",
    "                      \"_\"+\"blastp_score_split_list_S\"\n",
    "                      + str(threshold_score)+\"_\"+str(parallel_num), 'r') as blastp_score:\n",
    "                for each_line in blastp_score:\n",
    "                    split_each_line = each_line.split(\" \")\n",
    "                    split_each_line[2] = int(split_each_line[2])\n",
    "                    split_each_line[3] = int(split_each_line[3])\n",
    "                    blastp_score_split_list.append(split_each_line)\n",
    "    else:\n",
    "        for parallel_num in range(cpu_count):\n",
    "            parallel_num += 1\n",
    "            with open(Score_file+selected_species_dic[species_of_query]+\"_\"\n",
    "                      + selected_species_dic[species_of_subject]+\"_\"+\"best_score_S\"\n",
    "                      + str(threshold_score)+\"_\"+str(parallel_num), \"r\") as best_hit_score:\n",
    "                for each_line in best_hit_score:\n",
    "                    split_each_line = each_line.split(\" \")\n",
    "\n",
    "                    split_each_line[3] = int(split_each_line[3])\n",
    "                    forward_best_hit_score_list.append(split_each_line)\n",
    "            with open(Score_file+selected_species_dic[species_of_query] + \"_\"\n",
    "                      + selected_species_dic[species_of_subject]\n",
    "                      + \"_\"+\"blastp_score_split_list_S\"\n",
    "                      + str(threshold_score)+\"_\"+str(parallel_num), 'r') as blastp_score:\n",
    "                for each_line in blastp_score:\n",
    "                    split_each_line = each_line.split(\" \")\n",
    "                    split_each_line[2] = int(split_each_line[2])\n",
    "                    split_each_line[3] = int(split_each_line[3])\n",
    "                    blastp_score_split_list.append(split_each_line)\n",
    "\n",
    "    # \"bar = Bar(\"Searching : \"+selected_species_dic[species_of_query]+\"-\"+selected_species_dic[species_of_subject], max = len(forward_best_hit_score_list))\" #remove \"\"\n",
    "\n",
    "    for forward_best_hit_score_element in forward_best_hit_score_list:\n",
    "        matching_list = []\n",
    "        backward_best_score = ['-1', '-1', '-1']\n",
    "        # bar.next()\n",
    "        for element in blastp_score_split_list:\n",
    "            if element[1] == forward_best_hit_score_element[1]:\n",
    "                matching_list.append(element)\n",
    "\n",
    "        for element in matching_list:\n",
    "            if int(element[2]) > int(backward_best_score[2]):\n",
    "                backward_best_score = element\n",
    "    #        with open('./'+selected_species_dic[species_of_query]\\\n",
    "    # +\"_\"+selected_species_dic[species_of_subject]+'_subtraction'+\"_\"+str(threshold_score), 'a') as subtraction :\n",
    "    #            save_data = int(backward_best_score[2]) - int(forward_best_hit_score_element[2])\n",
    "    #            subtraction.write(str(save_data)+\"\\n\")\n",
    "\n",
    "        if int(backward_best_score[2]) - int(forward_best_hit_score_element[2]) <= threshold_score:\n",
    "            with open(Score_file+selected_species_dic[species_of_query]\n",
    "                      + \"_\"+selected_species_dic[species_of_subject]\n",
    "                      + \"_oneway_threshold_best_hit_Score\"+str(threshold_score), \"a\")\\\n",
    "                    as other_oneway_threshold_best_hit:\n",
    "                save_data = forward_best_hit_score_element[0]\\\n",
    "                    + \" \"+forward_best_hit_score_element[1]\\\n",
    "                    + \" \" + str(int(forward_best_hit_score_element[2]))+\"\\n\"\n",
    "                other_oneway_threshold_best_hit.write(save_data)\n",
    "\n",
    "    # bar.finish()\n",
    "    finish_time_bbh = time.time()\n",
    "    rbh_time = float((finish_time_bbh - start_time_bbh)/60)\n",
    "    print(bck_info+ \" %s-%s took %.2f minutes\" %\n",
    "          (selected_species_dic[species_of_query], selected_species_dic[species_of_subject], rbh_time))\n",
    "    return rbh_time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 11. search_equal_bbh_data(target_A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "code_folding": [
     0
    ],
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "def search_unequal_bbh_data(target_b):\n",
    "    \"\"\"Search the unequal backward best hit data. ex) AAE_CAC_backward_best_hit\n",
    "    No Return !! vlaues means Null return\"\"\"\n",
    "    global unequal_BBH_data\n",
    "    print(unequal_BBH_data)\n",
    "    for i in unequal_BBH_data:\n",
    "        if i[2] == 0:\n",
    "            \" \"\n",
    "            pass\n",
    "        else:\n",
    "            if target_b[0] == i[0] or target_b[0] == i[1] or target_b[1] == i[0] or target_b[1] == i[1]:\n",
    "                # sample of B  - [\"AAE_gi|156|ref....\",\"CAC_gi|1560..\",85]\n",
    "                copy_i = copy.copy(i)\n",
    "                tasks.put(copy_i)\n",
    "                unequal_BBH_data[unequal_BBH_data.index(i)][2] = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "######  12. search_unequal_bbh_data(target_B):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def search_unequal_bbh_data(target_b):\n",
    "    \"\"\"Search the unequal backward best hit data. ex) AAE_CAC_backward_best_hit\n",
    "    No Return !! vlaues means Null return\"\"\"\n",
    "    global unequal_BBH_data\n",
    "    print(unequal_BBH_data)\n",
    "    for i in unequal_BBH_data:\n",
    "        if i[2] == 0:\n",
    "            \" \"\n",
    "            pass\n",
    "        else:\n",
    "            if target_b[0] == i[0] or target_b[0] == i[1] or target_b[1] == i[0] or target_b[1] == i[1]:\n",
    "                # sample of B  - [\"AAE_gi|156|ref....\",\"CAC_gi|1560..\",85]\n",
    "                copy_i = copy.copy(i)\n",
    "                tasks.put(copy_i)\n",
    "                unequal_BBH_data[unequal_BBH_data.index(i)][2] = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 13.matching_bbh(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def matching_bbh(target):\n",
    "    \"\"\" Match the backward best hit \"\"\"\n",
    "    #print(\"running matching_bbh\")\n",
    "    if target[2] == 0:\n",
    "        return\n",
    "\n",
    "    else:\n",
    "        copy_target = copy.copy(target)\n",
    "        search_equal_bbh_data(copy_target[0])\n",
    "        search_equal_bbh_data(copy_target[1])\n",
    "        results.put(copy_target)\n",
    "        unequal_BBH_data[unequal_BBH_data.index(target)][2] = 0\n",
    "\n",
    "    for j in unequal_BBH_data:\n",
    "        if j[2] == 0:\n",
    "            pass\n",
    "\n",
    "        else:\n",
    "            if copy_target[0] == j[0] or copy_target[0] == j[1] or copy_target[1] == j[0] or copy_target[1] == j[1]:\n",
    "                copy_j = copy.copy(j)\n",
    "                unequal_BBH_data[unequal_BBH_data.index(j)][2] = 0\n",
    "                tasks.put(copy_j)\n",
    "\n",
    "    while not tasks.empty():\n",
    "        get_task = tasks.get()\n",
    "        search_equal_bbh_data(get_task[0])\n",
    "        search_equal_bbh_data(get_task[1])\n",
    "        results.put(get_task)\n",
    "        search_unequal_bbh_data(get_task)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 14. matrix_clustering_ortholog(element_set) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def matrix_clustering_ortholog(element_set):\n",
    "    # matrix_clustering_ortholog(element_set, bar): # this is Old Method old is generating is  removed\n",
    "    \"\"\" Generate the matrix of clustering ortholog. \"\"\"\n",
    "    print(\"matrix_clustering_ortholog Function is running element_set is Passed is \")\n",
    "    print(element_set)\n",
    "    row_data = []\n",
    "    col_data = []\n",
    "    temp_results = queue.Queue()\n",
    "    \"Create a queue object with a given maximum size. If maxsize is <=0, teh queue size is infinite\"\n",
    "    # bar.next()\n",
    "    for element in element_set:\n",
    "        # if element[0] exist, returning the index in the row_data.\n",
    "        if row_data.count(element[0]) > 0:\n",
    "            # element[0] is data of row. ['gi|15606057|ref|NP_213434.1|', 'gi|15606057|ref|NP_213434.1|', '3823\\n']\n",
    "            row = row_data.index(element[0])\n",
    "\n",
    "        else:\n",
    "            row = len(row_data)\n",
    "            row_data.append(element[0])\n",
    "            # if element[0] doesn't exist, appending the element[0] to the col_data.\n",
    "            if col_data.count(element[0]) < 1:\n",
    "                col_data.append(element[0])\n",
    "\n",
    "        if col_data.count(element[1]) > 0:\n",
    "            col = col_data.index(element[1])  # element[1] is data of col.\n",
    "\n",
    "        else:\n",
    "            col = len(col_data)\n",
    "            col_data.append(element[1])\n",
    "            if row_data.count(element[1]) < 1:\n",
    "                col_data.append(element[1])\n",
    "\n",
    "        temp_results.put([row, col, element[2]])\n",
    "\n",
    "    # create a new matrix of given shape(the size_resuls) and type, filled with zeros.\n",
    "    score_matrix = numpy.matlib.zeros(\n",
    "        (len(row_data), len(col_data)), dtype=np.float)\n",
    "    # np.zeros() can be used, will test later,\n",
    "    while not temp_results.empty():\n",
    "        get_temp_results = temp_results.get()\n",
    "        row = get_temp_results[0]\n",
    "        col = get_temp_results[1]\n",
    "        score_matrix[row, col] = get_temp_results[2]\n",
    "        score_matrix[col, row] = get_temp_results[2]\n",
    "\n",
    "    # If the elements of row and col is less than 2, it is excluded.\n",
    "    if len(row_data)*len(col_data) > 4:\n",
    "        # The big size of matrix(bigger than 1000 X 1000) will be computed by parallel_matrix_multiplication function.\n",
    "        if len(row_data) > 1000 and cpu_count > 1:\n",
    "            score_matrix = parallel_mcl(score_matrix)\n",
    "        else:\n",
    "            score_matrix = mcl(score_matrix)\n",
    "        clustering(row_data, col_data, score_matrix)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 15.parallel_mcl(score_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def parallel_mcl(score_matrix):\n",
    "    # print(\"parallel_mcl\")\n",
    "    count = 0\n",
    "    infinitesimal_value = 10**-10\n",
    "    idempotent_matrix = numpy.matlib.ones((2, 2))\n",
    "\n",
    "    while idempotent_matrix.sum() > infinitesimal_value:  # > infinitesimal_value\n",
    "        mcl_time_start = time.time()\n",
    "        pool = multiprocessing.Pool(cpu_count)  # create a expansion_matrix\n",
    "        multiplication_results = pool.map(parallel_matrix_multiplication,\n",
    "                                          zip(score_matrix, repeat(score_matrix)))\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "\n",
    "        # create a inflation_matrix(part 1)\n",
    "        pool = multiprocessing.Pool(cpu_count)\n",
    "        power_results = pool.map(\n",
    "            parallel_matrix_power, multiplication_results)\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "\n",
    "        sum_matrix = 0\n",
    "        for i in power_results:\n",
    "            sum_matrix = i + sum_matrix\n",
    "\n",
    "        # create a inflation_matrix(part 2)\n",
    "        pool = multiprocessing.Pool(cpu_count)\n",
    "        divide_results = pool.map(parallel_matrix_division,\n",
    "                                  zip(power_results, repeat(sum_matrix)))\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "\n",
    "        # Make a Combined matrix for results of parallel_matrix_multiplication function.\n",
    "        for i in range(len(divide_results)):\n",
    "            if i == 0:\n",
    "                score_matrix = divide_results[i]\n",
    "            else:\n",
    "                score_matrix = np.concatenate(\n",
    "                    (score_matrix, divide_results[i]), axis=0)\n",
    "\n",
    "        sum_results = 0\n",
    "        for i in multiplication_results:\n",
    "            sum_results += i\n",
    "        # identify whether inflation_matrix is idempotent matrix or not.\n",
    "        idempotent_matrix = abs(np.sum(score_matrix) - sum_results)\n",
    "\n",
    "        count += 1\n",
    "        if count > infinite_loop:  # It will prevent the infinite loop of mcl algorithm.\n",
    "            break\n",
    "        mcl_time_finish = time.time()\n",
    "        if verbose:\n",
    "            print(\" mcl time : %f, count : %d, matrix size : %d * %d\"\n",
    "                  % ((mcl_time_finish - mcl_time_start)/60, count, score_matrix[0].size, score_matrix[0].size))\n",
    "    return score_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 16. mcl(score_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mcl(score_matrix):\n",
    "    count = 0\n",
    "    infinitesimal_value = 10**-10\n",
    "    idempotent_matrix = numpy.matlib.ones((2, 2))\n",
    "    #\"idempotent_matrix = np.ones((2,2)) # i will later test with this\"\n",
    "    while idempotent_matrix.sum() > infinitesimal_value:  # > infinitesimal_value\n",
    "        mcl_time_start = time.time()\n",
    "        expansion_matrix = score_matrix ** 2\n",
    "        print(expansion_matrix)\n",
    "        print(\"\\n\")\n",
    "        print(inflation_factor)\n",
    "        score_matrix = np.power(expansion_matrix, inflation_factor)\n",
    "        score_matrix_sum = score_matrix.sum(axis=0)\n",
    "        # create a inflation_matrix\n",
    "        score_matrix = np.divide(score_matrix, score_matrix_sum)\n",
    "        # identify whether inflation_matrix is idempotent matrix or not.\n",
    "        idempotent_matrix = abs(score_matrix - expansion_matrix)\n",
    "        count += 1\n",
    "        if count > infinite_loop:  # It will prevent the infinite loop of mcl algorithm.\n",
    "            break\n",
    "        mcl_time_finish = time.time()\n",
    "        if verbose:\n",
    "            print(\" mcl time : %f, count : %d, matrix size : %d * %d\"\n",
    "                  % ((mcl_time_finish - mcl_time_start)/60, count, score_matrix[0].size, score_matrix[0].size))\n",
    "    return score_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 17. clustering(row_data, col_data, score_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def clustering(row_data, col_data, score_matrix):\n",
    "    global cluster_count\n",
    "    global ortholog_count\n",
    "    ortholog_temp_list = []\n",
    "    for i in range(len(row_data)):\n",
    "        ortholog_list = []\n",
    "        ortholog = queue.Queue()  # It is Queue which is put the ortholog.\n",
    "        # It is Queue which is put the ortholog having changed gene ID\n",
    "        gene_id_queue = queue.Queue()\n",
    "        for j in range(len(col_data)):\n",
    "            if 0.1 <= score_matrix[i, j]:\n",
    "                ortholog.put(col_data[j])\n",
    "                ortholog_list.append(col_data[j])\n",
    "\n",
    "        # If the ortholog queue  has the element of ortholog more than 2, it will be printed.\n",
    "        if ortholog.qsize() >= 3:\n",
    "            for element in ortholog_list:\n",
    "                try:\n",
    "                    ortholog_sum += ortholog_temp_list.index(element)+1\n",
    "\n",
    "                except ValueError:\n",
    "                    with open(Cluster_out+\"_geneID_S\"+str(threshold_score) +\n",
    "                              \"_\"+str(inflation_factor), \"a\") as ortholog_list_save:\n",
    "                        ortholog_print = \"cluster \"+str(cluster_count)+\" :\"\n",
    "                        ortholog_list_save.write(ortholog_print)\n",
    "\n",
    "                        while not ortholog.empty():\n",
    "                            get_ortholog = ortholog.get()\n",
    "                            ortholog_list_save.write(\"\\t\"+get_ortholog)\n",
    "\n",
    "                            try:\n",
    "                                # get_ortholog --> ECO_170082288,  get_ortholog.split('_') --> ['ECO', '170082288']\n",
    "                                get_ortholog_split = get_ortholog.split('_')\n",
    "                                gene_id_queue.put(\n",
    "                                    get_ortholog_split[0]+\"_\"+gene_id_dic[get_ortholog_split[1]])\n",
    "                            except KeyError:\n",
    "                                # If the gene_id_dic don't have get_ortholog, it will print the original ID(get_ortholog).\n",
    "                                gene_id_queue.put(get_ortholog)\n",
    "                        ortholog_list_save.write(\"\\n\")\n",
    "                    with open(Cluster_out+\"_KO_ID_S\"\n",
    "                              + str(threshold_score)\n",
    "                              + \"_\"+str(inflation_factor), \"a\")\\\n",
    "                            as ortholog_list_geneID_save:\n",
    "                        ortholog_list_geneid_print = \"cluster \" + \\\n",
    "                            str(cluster_count)+\" :\"\n",
    "                        ortholog_list_geneID_save.write(\n",
    "                            ortholog_list_geneid_print)\n",
    "                        while not gene_id_queue.empty():\n",
    "                            ortholog_list_geneID_save.write(\n",
    "                                \"\\t\"+gene_id_queue.get())\n",
    "                            ortholog_count += 1\n",
    "                        cluster_count += 1\n",
    "                        ortholog_list_geneID_save.write(\"\\n\")\n",
    "                    break\n",
    "            ortholog_temp_list = operator.concat(\n",
    "                ortholog_temp_list, ortholog_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 18.parallel_matrix_multiplication(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def parallel_matrix_multiplication(data):\n",
    "    \"Doing parallel_matrix_multiplication Using Numpy\"\n",
    "    matrix_element, matrix = data\n",
    "    return  matrix_element * matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 19.parallel_matrix_power(matrix_element):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def parallel_matrix_power(matrix_element):\n",
    "    \"This Function Compute Parallel_Matrix Power by using Numpy np.power() Function\"\n",
    "    global inflation_factor\n",
    "    return np.power(matrix_element, inflation_factor)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 20.parallel_matrix_division(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def parallel_matrix_division(data):\n",
    "    \"This Function Will Perform Matrix Division by Using Numpy library\"\n",
    "    matrix_element, sum_data = data\n",
    "    return np.divide(matrix_element, sum_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 21. read_species(pr = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "code_folding": [
     0
    ],
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": [
    "def read_species(pr=1):  # default 1 which will always shows the name of Species\n",
    "    \"\"\" If pr is 1, it will print \"Species_List\"  Other wise only return the Value in dic Format \"\"\"\n",
    "    read_species = os.listdir(Species)\n",
    "    selected_species_dic = {}  # list\n",
    "    backward_selected_species_dic = {}  # list\n",
    "    for i, species in enumerate(sorted(read_species), start=1):\n",
    "        selected_species_dic[i] = species\n",
    "        backward_selected_species_dic[species] = i\n",
    "        if pr == 1:\n",
    "            print(str(i)+\".\", species)\n",
    "        number = i\n",
    "    return selected_species_dic, backward_selected_species_dic, number"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 22. Del_File(path,file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "code_folding": [],
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "def del_file(path, file):\n",
    "    \"This Function delete the file passed with path and file\"\n",
    " #   \"Delete the Unnecessary file according to path and file name passed\"\n",
    " #  del_file =subprocess.Popen([\"rm \"+path+file], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)\n",
    " #   del_file_stream = del_file.communicate()\n",
    " #   if not del_file_stream[1]:\n",
    " #       print (\"Done to del \"+path+file)\n",
    " #   elif del_file_stream[1]:\n",
    " #       print (del_file_stream[1])\n",
    "    try:\n",
    "        os.remove(path+file)\n",
    "        print(\"File Successfully Removed\")\n",
    "    except:\n",
    "        print(\"Check the File or Path\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Del_File(\"./query/\",\"query_2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 23.check_file(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "code_folding": [
     0
    ],
    "run_control": {
     "marked": false
    }
   },
   "outputs": [],
   "source": [
    "def check_file(file):\n",
    "    \"\"\"When mode 3 is Selected.This function will run. file is user input value for clustering output.\n",
    "    Check the file weather exists or not . if exist this will warn to use other name\n",
    "    Cluster_out is file.Cluster_out(results) + _geneID_S+ str(threshold_score)+\"_\"\\\n",
    "        +str(inflation_factor) or Cluster_out+\"_KO_ID_S\"+str(threshold_score)+\"_\"+str(inflation_factor))\n",
    "        are files created by mcl algorithm so if these file exist system will exit\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"\n",
    "    Python2 Code\n",
    "        file_list = glob.glob(file+'*')    \n",
    "    if (Cluster_out+\"_geneID_S\"+str(threshold_score)+\"_\"+str(inflation_factor) or Cluster_out+\"_KO_ID_S\"+str(threshold_score)+\"_\"+str(inflation_factor)) in file_list:\n",
    "        print \"Please, set other name of output.\"\n",
    "        sys.exit(2)\n",
    "    \"\"\"\n",
    "    # Declare all variable as a globally Added\n",
    "    global Cluster_out, threshold_score, infinite_loop\n",
    "    file_list = glob.glob(file+'*')  # list all related files in same directory local variable\n",
    "    if (Cluster_out+\"_geneID_S\"+str(threshold_score)\n",
    "        + \"_\"+str(inflation_factor) or Cluster_out+\"_KO_ID_S\"\n",
    "            + str(threshold_score)+\"_\"+str(inflation_factor)) in file_list:\n",
    "        print(\"Please, set other name of output.line Number 966\")\n",
    "        sys.exit(2)\n",
    "    else :\n",
    "        print(file,\" is checked\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 24.read_equal_bbh(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def read_equal_bbh(path):\n",
    "    \"\"\"Read_Equal BBH by user path Blast Best Hit\n",
    "    the score fiie location is ./score_file/ + species + species like ./score_file/A_C\n",
    "    \"\"\"\n",
    "    global threshold_score\n",
    "\n",
    "    with open(path+\"_oneway_threshold_best_hit_Score\"\n",
    "              + str(threshold_score), 'r') as equal_RBH:\n",
    "        for j in equal_RBH:\n",
    "            split_data = j.split()\n",
    "            # Split Data sample\n",
    "            # ['A_gi|15605613|ref|NP_212986.1|', 'C_gi|15004707|ref|NP_149167.1|', '51']\n",
    "            split_data[2] = int(split_data[2]) #Only Converting to int format\n",
    "            equal_BBH_data.append(split_data)\n",
    "            equal_BBH_data_dic[split_data[0]] = split_data[1:]\n",
    "    try:\n",
    "        with open(path+\"_second_oneway_threshold_best_hit_Score\"\n",
    "                  + str(threshold_score), 'r') as second_equal_RBH:\n",
    "            for j in second_equal_RBH:\n",
    "                split_data = j.split()\n",
    "                #\"['AAE_gi|15606128|ref|NP_213505.1|', 'AAE_gi|15606877|ref|NP_214257.1|', 898] split_data sample\"\n",
    "                split_data[2] = int(split_data[2])\n",
    "                second_equal_BBH_data.append(split_data)\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### 25. read_unequal_bbh(path):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def read_unequal_bbh(path):\n",
    "    \"\"\" Read unequal BBH path passed by User.\n",
    "    Since this Function will append value to unequal_BBH_data which created before running this function\n",
    "    Later i will changed to return type of list from this function.create blank list and append and return\"\"\"\n",
    "\n",
    "    with open(path+\"_oneway_threshold_best_hit_Score\"\n",
    "              + str(threshold_score), 'r') as unequal_RBH:\n",
    "        for j in unequal_RBH:\n",
    "            split_data = j.split()\n",
    "            split_data[2] = int(split_data[2])\n",
    "            unequal_BBH_data.append(split_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read_unequal_bbh(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  Main Program Testing\n",
    "`************* No Function below **************************`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> Select a mode or modes (1 2 *OR* 1 3 *OR* 2 3): 1\n",
      ">> Select Genomes to detect Orthologs(e.g. 1 2 3 4 5 or 1-5) : 1 2\n",
      "\n",
      "Enter a matrix number: 2\n",
      "You can use 12 Core.\n",
      "If you input >= 2,    The Program will run a parallel computation for the blastp.\n",
      "Enter the number of Core to use in this program (1 ~ 12): 5\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(\"1. BLASTP. \\n2. BLASTP using precalculated data. \\n3. clustering.\\n\")\n",
    "mode = input(\">> Select a mode or modes (1 2 *OR* 1 3 *OR* 2 3): \")\n",
    "selected_species_dic, backward_selected_species_dic, number_i = read_species()\n",
    "selected_number = input(\n",
    "    \">> Select Genomes to detect Orthologs(e.g. 1 2 3 4 5 or 1-5) : \")\n",
    "\n",
    "if selected_number.find('-') > 0:\n",
    "    # find() return the index position of first occurance\n",
    "    SN = selected_number.split(\"-\")\n",
    "    if int(SN[-1]) > number_i:  # number_i is length of Genome file inside folder exit the process\n",
    "        print(\"\\nWrongInput\\nInput must be less than\", number_i)\n",
    "        sys.exit(2)\n",
    "    else:\n",
    "        user_selected_number = range(int(SN[0]), int(SN[-1])+1)\n",
    "        for j in user_selected_number:\n",
    "            print(selected_species_dic[j], end=\" \")  # loop in Dic\n",
    "        print(\"Selected!!\")\n",
    "\n",
    "else:\n",
    "    user_selected_number = sorted(\n",
    "        set([int(read_species) for read_species in selected_number.split()]))\n",
    "    # Create a set (remove repeating)\n",
    "    if int(user_selected_number[-1]) > number_i:\n",
    "        print(\"\\nWrongInput\\nInput must be less than\", number_i)\n",
    "        sys.exit(2)\n",
    "        # Greater than Genome list will system error\n",
    "    else:\n",
    "        for j in user_selected_number:\n",
    "            print(selected_species_dic[j], end=\" \")\n",
    "        print(\"Selected!!\")\n",
    "blastp_matrix = matrix_name()\n",
    "cpu_count = int(input(\"You can use %s Core.\\nIf you input >= 2,\\\n",
    "    The Program will run a parallel computation for the blastp.\\n\" % multiprocessing.cpu_count()\n",
    "                          + \"Enter the number of Core to use in this program (1 ~ %s): \" % multiprocessing.cpu_count()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### part 2  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "if \"3\" in mode:\n",
    "    \"3 is for BLOSUM82\"\n",
    "    inflation_factor = input(\"inflation factor to cluster\")\n",
    "    Cluster_out= input(\"Set the name of clustering output Folder\")\n",
    "    \"cluster_out is user  input value\"\n",
    "    check_file(Cluster_out)\n",
    "    \n",
    "    Log_file_name = \"./cluster_out/\"+Cluster_out+\"_S\"+str(threshold_score)+\"_\"+str(inflation_factor)+\".log\"\n",
    "    with open(Log_file_name+datetime.datetime.now().strftime(\"_%Y_%m_%d_%H_%M_%S\")+\".txt\", 'w') as log:\n",
    "        \"if 3 is passed in mode then the log file created with Cluster and inflation_factor\"\n",
    "        log.write(\"\\t\\t\\tThis Log File is Created on\\t\")\n",
    "        log.write(str(datetime.datetime.now()))\n",
    "        log.write(\"\\nmode :\")\n",
    "        for i in mode:\n",
    "            log.write(\" \"+i)\n",
    "        log.write(\"\\ngenomes : \")\n",
    "        for i in user_selected_number:\n",
    "            log.write(selected_species_dic[i]+\" \")\n",
    "        log.write(\"\\ncpu_count : \"+str(cpu_count))\n",
    "        log.write(\"\\nblastp matrix : \"+blastp_matrix)\n",
    "        if \"3\" in mode:\n",
    "            log.write(\"\\ninflation_factor : \"+str(inflation_factor))\n",
    "            log.write(\"\\nCluster out : \"+Cluster_out)\n",
    "        log.write(\"\\nSpecies : \"+Species)\n",
    "        log.write(\"\\nBlastp : \"+Blastp)\n",
    "        log.write(\"\\nScore file : \"+Score_file)\n",
    "        log.write(\"\\nBlastp_data : \"+Blastp_data)\n",
    "        log.write(\"\\nsave rawblastp score : \"+str(save_raw_blastp_score))\n",
    "        log.write(\"\\n\")\n",
    "elif not \"3\" in mode:\n",
    "    \"\"\"If mode 3 Clustering is not Passed then simple log wile will created \"\"\"\n",
    "    Log_file_name = 'log_files/Log'  # log file rename + .txt\n",
    "    with open(Log_file_name+datetime.datetime.now().strftime(\"_%Y_%m_%d_%H_%M_%S\")+\".txt\", 'w') as log:\n",
    "        log.write(\"\\t\\t\\tThis Log File is Created on\\t\")\n",
    "        log.write(str(datetime.datetime.now()))\n",
    "        log.write(\"\\nmode :\")\n",
    "        for i in mode:\n",
    "            log.write(\" \"+i)\n",
    "        log.write(\"\\ngenomes : \")\n",
    "        for i in user_selected_number:\n",
    "            log.write(selected_species_dic[i]+\" \")\n",
    "        log.write(\"\\ncpu_count : \"+str(cpu_count))\n",
    "        log.write(\"\\nblastp matrix : \"+blastp_matrix)\n",
    "        if \"3\" in mode:\n",
    "            log.write(\"\\ninflation_factor : \"+str(inflation_factor))\n",
    "            log.write(\"\\nCluster out : \"+Cluster_out)\n",
    "        log.write(\"\\nSpecies : \"+Species)\n",
    "        log.write(\"\\nBlastp : \"+Blastp)\n",
    "        log.write(\"\\nScore file : \"+Score_file)\n",
    "        log.write(\"\\nBlastp_data : \"+Blastp_data)\n",
    "        log.write(\"\\nsave rawblastp score : \"+str(save_raw_blastp_score))\n",
    "        log.write(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running the blastp & forward best hit searches  between AAE genome and AAE genome\n"
     ]
    },
    {
     "ename": "UnicodeEncodeError",
     "evalue": "'cp949' codec can't encode character '\\u26a0' in position 1: illegal multibyte sequence",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\alive_progress\\progress.py\u001b[0m in \u001b[0;36malive_bar\u001b[1;34m(total, title, calibrate, **options)\u001b[0m\n\u001b[0;32m    264\u001b[0m     \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 265\u001b[1;33m         \u001b[1;32myield\u001b[0m \u001b[0mbar\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    266\u001b[0m     \u001b[1;32mfinally\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-12-11e03ed0f8d8>\u001b[0m in \u001b[0;36mrun_parallel_query\u001b[1;34m(species_of_query, species_of_subject, query_v, parallel_num)\u001b[0m\n\u001b[0;32m     21\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mblastp_score\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m:\u001b[0m  \u001b[1;31m# if blastp run successfully\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 22\u001b[1;33m                 best_score, blastp_score_split_list = forward_best_hit(\n\u001b[0m\u001b[0;32m     23\u001b[0m                     blastp_score.decode())  # .decode() Convert in to str format\n",
      "\u001b[1;31mNameError\u001b[0m: name 'forward_best_hit' is not defined",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mUnicodeEncodeError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-34-c4c8578942af>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0mstart_time_OBH\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;32mif\u001b[0m \u001b[1;34m\"1\"\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mmode\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m     \u001b[0mbackward_best_hit_work_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0moneway_threshold_best_hit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmode\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m     \u001b[0mpool\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmultiprocessing\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mPool\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcpu_count\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[0mresults\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpool\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mbackward_best_hit\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mbackward_best_hit_work_list\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-13-278e574d8e31>\u001b[0m in \u001b[0;36moneway_threshold_best_hit\u001b[1;34m(mode)\u001b[0m\n\u001b[0;32m     24\u001b[0m                         \u001b[1;31m#\"No Parallel Computing while cpu count == 1\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m                         \u001b[0mblastp_time_start\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m                         \u001b[0mrun_parallel_query\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mquery_v\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcpu_count\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     27\u001b[0m                         \u001b[1;31m#\"i is first species k is second species number queryv is list file of i Position genome\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     28\u001b[0m                         \u001b[0mblastp_time_end\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-12-11e03ed0f8d8>\u001b[0m in \u001b[0;36mrun_parallel_query\u001b[1;34m(species_of_query, species_of_subject, query_v, parallel_num)\u001b[0m\n\u001b[0;32m     83\u001b[0m                           \u001b[1;33m+\u001b[0m \u001b[1;34m\"_\"\u001b[0m\u001b[1;33m+\u001b[0m\u001b[0mselected_species_dic\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mspecies_of_subject\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;34m\"_S\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     84\u001b[0m                           + str(threshold_score)+\"_\"+str(parallel_num), \"a\") as save_blastp:\n\u001b[1;32m---> 85\u001b[1;33m                     \u001b[0msave_blastp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mblastp_score\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     86\u001b[0m     \u001b[1;31m# bar.finish() # progressing bar finish\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     87\u001b[0m     \u001b[1;32mreturn\u001b[0m  \u001b[1;31m# None\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\contextlib.py\u001b[0m in \u001b[0;36m__exit__\u001b[1;34m(self, type, value, traceback)\u001b[0m\n\u001b[0;32m    128\u001b[0m                 \u001b[0mvalue\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    129\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 130\u001b[1;33m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mthrow\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtype\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtraceback\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    131\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    132\u001b[0m                 \u001b[1;31m# Suppress StopIteration *unless* it's the same exception that\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\alive_progress\\progress.py\u001b[0m in \u001b[0;36malive_bar\u001b[1;34m(total, title, calibrate, **options)\u001b[0m\n\u001b[0;32m    273\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    274\u001b[0m         \u001b[0mend\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrun\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrun\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstats\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstats_end\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 275\u001b[1;33m         \u001b[0malive_repr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\anaconda3\\lib\\site-packages\\alive_progress\\progress.py\u001b[0m in \u001b[0;36malive_repr\u001b[1;34m(spin)\u001b[0m\n\u001b[0;32m    116\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mline_len\u001b[0m \u001b[1;33m<\u001b[0m \u001b[0mrun\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlast_line_len\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    117\u001b[0m                 \u001b[0mclear_traces\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 118\u001b[1;33m             \u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__stdout__\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mline\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mspin\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;34m'\\r'\u001b[0m \u001b[1;32mor\u001b[0m \u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    119\u001b[0m             \u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__stdout__\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    120\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mUnicodeEncodeError\u001b[0m: 'cp949' codec can't encode character '\\u26a0' in position 1: illegal multibyte sequence"
     ]
    }
   ],
   "source": [
    "start_time_OBH = time.time()\n",
    "if \"1\" in mode:\n",
    "    backward_best_hit_work_list = oneway_threshold_best_hit(mode)\n",
    "    pool = multiprocessing.Pool(cpu_count)\n",
    "    results = pool.map(backward_best_hit,backward_best_hit_work_list)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "elif \"2\" in mode:\n",
    "    used_precalculated_data_list = []\n",
    "    new_calculated_data_list = []\n",
    "    print(\"Threshold_score passed is \", threshold_score)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Python 2 whole Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precalculated_data_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id_dic = {}\n",
    "with open(\"./db/myva=gb\",\"r\")  as id_read:\n",
    "    for i in id_read:\n",
    "        gene_name, gene_id = i.split()\n",
    "        gene_id_dic[gene_id.replace(\"\\n\",\"\")]  = gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'14600509': 'APE0180',\n",
       " '14600543': 'APE0225',\n",
       " '14600591': 'APE0277',\n",
       " '14600619': 'APE0307',\n",
       " '14600631': 'APE0324',\n",
       " '14600634': 'APE0328',\n",
       " '14600637': 'APE0331',\n",
       " '14600648': 'APE0344',\n",
       " '14600665': 'APE0369',\n",
       " '14600671': 'APE0380',\n",
       " '14600704': 'APE0432',\n",
       " '14600725': 'APE0455',\n",
       " '14600744': 'APE0479',\n",
       " '14600745': 'APE0480',\n",
       " '14600789': 'APE0539',\n",
       " '14600799': 'APE0552',\n",
       " '14600819': 'APE0590',\n",
       " '14600848': 'APE0622',\n",
       " '14600853': 'APE0629',\n",
       " '14600865': 'APE0649',\n",
       " '14600886': 'APE0679',\n",
       " '14600943': 'APE0755',\n",
       " '14600950': 'APE0765',\n",
       " '14600954': 'APE0770',\n",
       " '14600975': 'APE0795',\n",
       " '14600987': 'APE0814',\n",
       " '14601075': 'APE0926',\n",
       " '14601100': 'APE0956',\n",
       " '14601137': 'APE1010',\n",
       " '14601142': 'APE1022',\n",
       " '14601144': 'APE1024',\n",
       " '14601155': 'APE1046',\n",
       " '14601159': 'APE1051',\n",
       " '14601170': 'APE1077',\n",
       " '14601177': 'APE1089',\n",
       " '14601182': 'APE1097',\n",
       " '14601192': 'APE1122',\n",
       " '14601198': 'APE1128',\n",
       " '14601214': 'APE1153',\n",
       " '14601248': 'APE1203',\n",
       " '14601259': 'APE1214',\n",
       " '14601274': 'APE1233',\n",
       " '14601297': 'APE1264',\n",
       " '14601321': 'APE1302',\n",
       " '14601349': 'APE1349',\n",
       " '14601380': 'APE1394',\n",
       " '14601401': 'APE1429',\n",
       " '14601437': 'APE1490',\n",
       " '14601449': 'APE1508',\n",
       " '14601454': 'APE1515',\n",
       " '14601481': 'APE1556',\n",
       " '14601552': 'APE1680',\n",
       " '14601598': 'APE1745',\n",
       " '14601636': 'APE1803',\n",
       " '14601696': 'APE1884',\n",
       " '14601698': 'APE1886',\n",
       " '14601718': 'APE1916',\n",
       " '14601739': 'APE1951',\n",
       " '14601763': 'APE1982',\n",
       " '14601764': 'APE1983',\n",
       " '14601775': 'APE2000',\n",
       " '14601805': 'APE2051',\n",
       " '14601840': 'APE2095',\n",
       " '14601890': 'APE2172',\n",
       " '14601918': 'APE2215',\n",
       " '14601926': 'APE2226',\n",
       " '14601990': 'APE2321',\n",
       " '14601996': 'APE2327',\n",
       " '14602011': 'APE2347',\n",
       " '14602025': 'APE2372',\n",
       " '14602035': 'APE2391',\n",
       " '14602057': 'APE2428',\n",
       " '14602075': 'APE2457',\n",
       " '14602109': 'APE2506',\n",
       " '14602124': 'APE2529',\n",
       " '14602128': 'APE2536',\n",
       " '14602132': 'APE2539',\n",
       " '14602167': 'APE2597',\n",
       " '14602176': 'APE2612',\n",
       " '14600380': 'APE0001',\n",
       " '14600381': 'APE0002',\n",
       " '14600382': 'APE0004',\n",
       " '14600383': 'APE0006',\n",
       " '14600384': 'APE0007',\n",
       " '14600385': 'APE0009',\n",
       " '14600386': 'APE0010',\n",
       " '14600387': 'APE0011',\n",
       " '14600388': 'APE0012',\n",
       " '14600389': 'APE0013',\n",
       " '14600390': 'APE0016',\n",
       " '14600391': 'APE0017',\n",
       " '14600392': 'APE0020',\n",
       " '14600393': 'APE0023',\n",
       " '14600394': 'APE0024',\n",
       " '14600395': 'APE0025',\n",
       " '14600396': 'APE0026',\n",
       " '14600397': 'APE0028',\n",
       " '14600398': 'APE0031',\n",
       " '14600399': 'APE0033',\n",
       " '14600400': 'APE0035',\n",
       " '14600401': 'APE0038',\n",
       " '14600402': 'APE0040',\n",
       " '14600403': 'APE0043',\n",
       " '14600404': 'APE0045',\n",
       " '14600405': 'APE0047',\n",
       " '14600406': 'APE0048',\n",
       " '14600407': 'APE0050',\n",
       " '14600408': 'APE0051',\n",
       " '14600409': 'APE0052',\n",
       " '14600410': 'APE0053',\n",
       " '14600411': 'APE0054',\n",
       " '14600412': 'APE0055',\n",
       " '14600413': 'APE0056',\n",
       " '14600414': 'APE0057',\n",
       " '14600415': 'APE0058',\n",
       " '14600416': 'APE0059',\n",
       " '14600417': 'APE0060',\n",
       " '14600418': 'APE0061',\n",
       " '14600419': 'APE0063',\n",
       " '14600420': 'APE0065',\n",
       " '14600421': 'APE0067',\n",
       " '14600422': 'APE0068',\n",
       " '14600423': 'APE0069',\n",
       " '14600424': 'APE0070',\n",
       " '14600425': 'APE0071',\n",
       " '14600426': 'APE0072',\n",
       " '14600427': 'APE0073',\n",
       " '14600428': 'APE0074',\n",
       " '14600429': 'APE0075',\n",
       " '14600430': 'APE0076',\n",
       " '14600431': 'APE0077',\n",
       " '14600432': 'APE0079',\n",
       " '14600433': 'APE0080',\n",
       " '14600434': 'APE0082',\n",
       " '14600435': 'APE0083',\n",
       " '14600436': 'APE0084',\n",
       " '14600437': 'APE0085',\n",
       " '14600438': 'APE0086',\n",
       " '14600439': 'APE0087',\n",
       " '14600440': 'APE0088',\n",
       " '14600441': 'APE0089',\n",
       " '14600442': 'APE0091',\n",
       " '14602183': 'APE0092m',\n",
       " '14600443': 'APE0093',\n",
       " '14600444': 'APE0094',\n",
       " '14600445': 'APE0095',\n",
       " '14600446': 'APE0096',\n",
       " '14600447': 'APE0097',\n",
       " '14600448': 'APE0099',\n",
       " '14600449': 'APE0101',\n",
       " '14600450': 'APE0103',\n",
       " '14600451': 'APE0104',\n",
       " '14600452': 'APES001',\n",
       " '14600453': 'APE0105',\n",
       " '14600454': 'APE0106',\n",
       " '14600455': 'APE0107',\n",
       " '14600456': 'APE0109',\n",
       " '14600457': 'APE0110',\n",
       " '14600458': 'APE0113',\n",
       " '14600459': 'APES002',\n",
       " '14600460': 'APE0115',\n",
       " '14600461': 'APE0116',\n",
       " '14600462': 'APE0117',\n",
       " '14600463': 'APE0119',\n",
       " '14600464': 'APE0121',\n",
       " '14600465': 'APE0122',\n",
       " '14600466': 'APE0124',\n",
       " '14600467': 'APE0125',\n",
       " '14600468': 'APE0126',\n",
       " '14600469': 'APE0127',\n",
       " '14600470': 'APE0129',\n",
       " '14600471': 'APE0130',\n",
       " '14600472': 'APE0131',\n",
       " '14600473': 'APE0132',\n",
       " '14600474': 'APE0134',\n",
       " '14600475': 'APE0135',\n",
       " '14600476': 'APE0136',\n",
       " '14600477': 'APE0137',\n",
       " '14600478': 'APE0140',\n",
       " '14600479': 'APE0141',\n",
       " '14600480': 'APE0142',\n",
       " '14600481': 'APE0143',\n",
       " '14600482': 'APE0145',\n",
       " '14600483': 'APE0146',\n",
       " '14600484': 'APE0147',\n",
       " '14600485': 'APE0148',\n",
       " '14600486': 'APE0150',\n",
       " '14600487': 'APE0151',\n",
       " '14600488': 'APE0152',\n",
       " '14602184': 'APE0152a',\n",
       " '14600489': 'APE0153',\n",
       " '14600490': 'APE0154',\n",
       " '14600491': 'APE0155',\n",
       " '14600492': 'APE0156',\n",
       " '14600493': 'APE0157',\n",
       " '14600494': 'APE0158',\n",
       " '14600495': 'APE0159',\n",
       " '14600496': 'APE0161',\n",
       " '14600497': 'APE0162',\n",
       " '14600498': 'APE0163',\n",
       " '14600499': 'APE0166',\n",
       " '14600500': 'APE0167',\n",
       " '14600501': 'APE0168',\n",
       " '14600502': 'APE0169',\n",
       " '14600503': 'APE0171',\n",
       " '14600504': 'APE0173',\n",
       " '14600505': 'APE0176',\n",
       " '14600506': 'APE0177',\n",
       " '14600507': 'APE0178',\n",
       " '14600508': 'APE0179',\n",
       " '14600510': 'APES003',\n",
       " '14600511': 'APE0181',\n",
       " '14600512': 'APE0182',\n",
       " '14600513': 'APE0183',\n",
       " '14600514': 'APE0184',\n",
       " '14600515': 'APE0186',\n",
       " '14600516': 'APE0187',\n",
       " '14600517': 'APE0188',\n",
       " '14600518': 'APE0191',\n",
       " '14602185': 'APE0193m',\n",
       " '14600519': 'APE0195',\n",
       " '14600520': 'APE0196',\n",
       " '14600521': 'APE0197',\n",
       " '14600522': 'APE0198',\n",
       " '14600523': 'APE0199',\n",
       " '14600524': 'APE0200',\n",
       " '14600525': 'APE0201',\n",
       " '14600526': 'APE0202',\n",
       " '14600527': 'APE0203',\n",
       " '14600528': 'APE0204',\n",
       " '14600529': 'APE0207',\n",
       " '14600530': 'APE0208',\n",
       " '14600531': 'APE0209',\n",
       " '14600532': 'APE0211',\n",
       " '14600533': 'APE0212',\n",
       " '14600534': 'APE0213',\n",
       " '14600535': 'APE0214',\n",
       " '14600536': 'APE0215',\n",
       " '14600537': 'APE0216',\n",
       " '14600538': 'APE0218',\n",
       " '14600539': 'APE0219',\n",
       " '14600540': 'APE0220',\n",
       " '14600541': 'APE0221',\n",
       " '14600542': 'APE0222',\n",
       " '14600544': 'APE0226',\n",
       " '14600545': 'APE0227',\n",
       " '14600546': 'APE0228',\n",
       " '14600547': 'APES004',\n",
       " '14600548': 'APE0230',\n",
       " '14600549': 'APE0231',\n",
       " '14600550': 'APE0233',\n",
       " '14600551': 'APE0234',\n",
       " '14600552': 'APE0235',\n",
       " '14600553': 'APE0236',\n",
       " '14600554': 'APE0237',\n",
       " '14600555': 'APE0238',\n",
       " '14600556': 'APE0239',\n",
       " '14600557': 'APE0240',\n",
       " '14600558': 'APE0241',\n",
       " '14600559': 'APE0242',\n",
       " '14600560': 'APE0243',\n",
       " '14600561': 'APE0244',\n",
       " '14600562': 'APE0246',\n",
       " '14600563': 'APE0248',\n",
       " '14600564': 'APE0249',\n",
       " '14600565': 'APE0252',\n",
       " '14600566': 'APE0254',\n",
       " '14600567': 'APE0255',\n",
       " '14600568': 'APE0256',\n",
       " '14600569': 'APES005',\n",
       " '14600570': 'APE0257',\n",
       " '14600571': 'APES006',\n",
       " '14600572': 'APE0258',\n",
       " '14600573': 'APE0260',\n",
       " '14600574': 'APE0261',\n",
       " '14600575': 'APE0262',\n",
       " '14600576': 'APE0263',\n",
       " '14600577': 'APE0264',\n",
       " '14600578': 'APE0265',\n",
       " '14600579': 'APE0266',\n",
       " '14600580': 'APE0267',\n",
       " '14600581': 'APE0269',\n",
       " '14600582': 'APES007',\n",
       " '14600583': 'APES008',\n",
       " '14600584': 'APE0270',\n",
       " '14600585': 'APE0272',\n",
       " '14600586': 'APE0273',\n",
       " '14600587': 'APE0274',\n",
       " '14600588': 'APE0275',\n",
       " '14600589': 'APES009',\n",
       " '14600590': 'APE0276',\n",
       " '14600592': 'APE0278',\n",
       " '14600593': 'APE0279',\n",
       " '14600594': 'APES010',\n",
       " '14600595': 'APE0280',\n",
       " '14600596': 'APE0282',\n",
       " '14600597': 'APE0283',\n",
       " '14600598': 'APE0284',\n",
       " '14600599': 'APE0285',\n",
       " '14600600': 'APES011',\n",
       " '14600601': 'APE0287',\n",
       " '14600602': 'APE0288',\n",
       " '14600603': 'APE0290',\n",
       " '14600604': 'APE0291',\n",
       " '14600605': 'APE0292',\n",
       " '14600606': 'APE0293',\n",
       " '14600607': 'APE0294',\n",
       " '14600608': 'APE0295',\n",
       " '14600609': 'APE0296',\n",
       " '14600610': 'APE0297',\n",
       " '14600611': 'APE0298',\n",
       " '14600612': 'APE0299',\n",
       " '14600613': 'APE0300',\n",
       " '14600614': 'APE0301',\n",
       " '14600615': 'APE0302',\n",
       " '14600616': 'APE0303',\n",
       " '14600617': 'APE0304',\n",
       " '14600618': 'APE0306',\n",
       " '14600620': 'APE0308',\n",
       " '14600621': 'APE0309',\n",
       " '14600622': 'APE0311',\n",
       " '14600623': 'APE0313',\n",
       " '14600624': 'APE0314',\n",
       " '14600625': 'APE0316',\n",
       " '14600626': 'APE0317',\n",
       " '14600627': 'APE0319',\n",
       " '14600628': 'APE0320',\n",
       " '14600629': 'APE0321',\n",
       " '14600630': 'APE0322',\n",
       " '14600632': 'APE0325',\n",
       " '14600633': 'APE0326',\n",
       " '14600635': 'APE0329',\n",
       " '14600636': 'APE0330',\n",
       " '14600638': 'APE0333',\n",
       " '14600639': 'APE0334',\n",
       " '14600640': 'APE0336',\n",
       " '14600641': 'APE0337',\n",
       " '14600642': 'APE0338',\n",
       " '14600643': 'APE0339',\n",
       " '14600644': 'APE0340',\n",
       " '14600645': 'APE0341',\n",
       " '14600646': 'APE0342',\n",
       " '14600647': 'APE0343',\n",
       " '14600649': 'APE0345',\n",
       " '14600650': 'APE0346',\n",
       " '14600651': 'APE0347',\n",
       " '14600652': 'APE0348',\n",
       " '14600653': 'APE0349',\n",
       " '14600654': 'APE0350',\n",
       " '14600655': 'APE0352',\n",
       " '14602186': 'APE0353a',\n",
       " '14600656': 'APE0354',\n",
       " '14600657': 'APE0356',\n",
       " '14600658': 'APE0358',\n",
       " '14600659': 'APE0359',\n",
       " '14600660': 'APE0360',\n",
       " '14600661': 'APE0362',\n",
       " '14602187': 'APE0362a',\n",
       " '14600662': 'APE0363',\n",
       " '14600663': 'APE0365',\n",
       " '14600664': 'APE0367',\n",
       " '14600666': 'APE0370',\n",
       " '14600667': 'APE0371',\n",
       " '14600668': 'APE0372',\n",
       " '14600669': 'APE0373',\n",
       " '14600670': 'APE0374',\n",
       " '14602188': 'APE0381m',\n",
       " '14600672': 'APE0382',\n",
       " '14600673': 'APE0383',\n",
       " '14600674': 'APE0384',\n",
       " '14600675': 'APE0385',\n",
       " '14600676': 'APE0389',\n",
       " '14600677': 'APE0390',\n",
       " '14600678': 'APE0392',\n",
       " '14600679': 'APE0396',\n",
       " '14600680': 'APE0399',\n",
       " '14600681': 'APE0401',\n",
       " '14600682': 'APES012',\n",
       " '14600683': 'APE0402',\n",
       " '14600684': 'APE0404',\n",
       " '14600685': 'APE0405',\n",
       " '14600686': 'APE0409',\n",
       " '14600687': 'APE0410',\n",
       " '14600688': 'APE0412',\n",
       " '14600689': 'APE0413',\n",
       " '14600690': 'APE0414',\n",
       " '14600691': 'APE0415',\n",
       " '14600692': 'APE0416',\n",
       " '14600693': 'APE0418',\n",
       " '14600694': 'APES014',\n",
       " '14600695': 'APE0419',\n",
       " '14600696': 'APE0421',\n",
       " '14600697': 'APE0423',\n",
       " '14600698': 'APE0424',\n",
       " '14600699': 'APE0426',\n",
       " '14600700': 'APE0427',\n",
       " '14600701': 'APE0428',\n",
       " '14600702': 'APE0429',\n",
       " '14600703': 'APE0431',\n",
       " '14600705': 'APE0433',\n",
       " '14602189': 'APE0433a',\n",
       " '14600706': 'APE0436',\n",
       " '14600707': 'APES015',\n",
       " '14600708': 'APES016',\n",
       " '14600709': 'APE0437',\n",
       " '14600710': 'APE0438',\n",
       " '14600711': 'APE0441',\n",
       " '14600712': 'APES017',\n",
       " '14600713': 'APE0442',\n",
       " '14600714': 'APES018',\n",
       " '14600715': 'APE0443',\n",
       " '14600716': 'APE0444',\n",
       " '14600717': 'APE0445',\n",
       " '14600718': 'APE0446',\n",
       " '14600719': 'APE0447',\n",
       " '14600720': 'APE0449',\n",
       " '14600721': 'APE0450',\n",
       " '14600722': 'APES019',\n",
       " '14600723': 'APE0452',\n",
       " '14600724': 'APE0454',\n",
       " '14600726': 'APE0457',\n",
       " '14600727': 'APE0459',\n",
       " '14600728': 'APE0460',\n",
       " '14600729': 'APE0461',\n",
       " '14600730': 'APE0462',\n",
       " '14600731': 'APE0464',\n",
       " '14600732': 'APE0465',\n",
       " '14600733': 'APE0468',\n",
       " '14600734': 'APE0469',\n",
       " '14600735': 'APE0470',\n",
       " '14600736': 'APE0471',\n",
       " '14600737': 'APES020',\n",
       " '14600738': 'APE0472',\n",
       " '14600739': 'APES021',\n",
       " '14600740': 'APE0473',\n",
       " '14600741': 'APE0475',\n",
       " '14600742': 'APE0476',\n",
       " '14600743': 'APE0478',\n",
       " '14600746': 'APE0481',\n",
       " '14600747': 'APE0482',\n",
       " '14600748': 'APE0483',\n",
       " '14600749': 'APE0485',\n",
       " '14600750': 'APE0486',\n",
       " '14600751': 'APE0487',\n",
       " '14600752': 'APE0488',\n",
       " '14600753': 'APE0489',\n",
       " '14600754': 'APE0490',\n",
       " '14600755': 'APE0492',\n",
       " '14600756': 'APE0493',\n",
       " '14600757': 'APE0494',\n",
       " '14600758': 'APE0495',\n",
       " '14600759': 'APE0496',\n",
       " '14600760': 'APE0498',\n",
       " '14600761': 'APE0500',\n",
       " '14600762': 'APE0502',\n",
       " '14600763': 'APE0503',\n",
       " '14600764': 'APE0505',\n",
       " '14600765': 'APE0506',\n",
       " '14600766': 'APE0507',\n",
       " '14600767': 'APE0508',\n",
       " '14600768': 'APE0509',\n",
       " '14600769': 'APE0510',\n",
       " '14600770': 'APE0511',\n",
       " '14600771': 'APE0513',\n",
       " '14600772': 'APE0514',\n",
       " '14600773': 'APE0516',\n",
       " '14600774': 'APE0517',\n",
       " '14600775': 'APE0519',\n",
       " '14600776': 'APE0521',\n",
       " '14600777': 'APE0522',\n",
       " '14600778': 'APE0525',\n",
       " '14600779': 'APES022',\n",
       " '14600780': 'APES023',\n",
       " '14600781': 'APE0526',\n",
       " '14600782': 'APE0528',\n",
       " '14600783': 'APE0531',\n",
       " '14600784': 'APE0532',\n",
       " '14600785': 'APE0534',\n",
       " '14600786': 'APE0535',\n",
       " '14600787': 'APE0536',\n",
       " '14600788': 'APE0537',\n",
       " '14600790': 'APES024',\n",
       " '14600791': 'APE0540',\n",
       " '14600792': 'APE0541',\n",
       " '14600793': 'APE0542',\n",
       " '14600794': 'APE0544',\n",
       " '14600795': 'APE0545',\n",
       " '14600796': 'APE0546',\n",
       " '14602190': 'APE0548a',\n",
       " '14600797': 'APE0549',\n",
       " '14600798': 'APE0550',\n",
       " '14600800': 'APE0553',\n",
       " '14600801': 'APE0554',\n",
       " '14600802': 'APE0555',\n",
       " '14600803': 'APE0557',\n",
       " '14600804': 'APE0558',\n",
       " '14600805': 'APE0560',\n",
       " '14600806': 'APE0561',\n",
       " '14600807': 'APE0562',\n",
       " '14602191': 'APE0563a',\n",
       " '14600808': 'APE0564',\n",
       " '14600809': 'APE0569',\n",
       " '14600810': 'APE0572',\n",
       " '14600811': 'APE0574',\n",
       " '14600812': 'APE0576',\n",
       " '14600813': 'APE0577',\n",
       " '14600814': 'APE0579',\n",
       " '14600815': 'APE0581',\n",
       " '14600816': 'APE0583',\n",
       " '14600817': 'APE0586',\n",
       " '14600818': 'APE0588',\n",
       " '14600820': 'APE0591',\n",
       " '14600821': 'APE0593',\n",
       " '14600822': 'APE0594',\n",
       " '14600823': 'APES025',\n",
       " '14600824': 'APE0595',\n",
       " '14600825': 'APE0596',\n",
       " '14600826': 'APE0597',\n",
       " '14600827': 'APE0598',\n",
       " '14600828': 'APE0600',\n",
       " '14600829': 'APE0601',\n",
       " '14600830': 'APE0602',\n",
       " '14600831': 'APE0603',\n",
       " '14600832': 'APE0605',\n",
       " '14600833': 'APE0606',\n",
       " '14600834': 'APE0607',\n",
       " '14600835': 'APE0609',\n",
       " '14600836': 'APE0610',\n",
       " '14600837': 'APE0611',\n",
       " '14600838': 'APE0612',\n",
       " '14600839': 'APE0613',\n",
       " '14600840': 'APE0614',\n",
       " '14600841': 'APE0615',\n",
       " '14600842': 'APE0616',\n",
       " '14600843': 'APE0617',\n",
       " '14600844': 'APE0618',\n",
       " '14600845': 'APE0619',\n",
       " '14600846': 'APE0620',\n",
       " '14602192': 'APE0620a',\n",
       " '14600847': 'APE0621',\n",
       " '14600849': 'APE0623',\n",
       " '14600850': 'APE0624',\n",
       " '14600851': 'APE0627',\n",
       " '14600852': 'APE0628',\n",
       " '14600854': 'APE0630',\n",
       " '14600855': 'APE0632',\n",
       " '14600856': 'APES027',\n",
       " '14600857': 'APE0636',\n",
       " '14600858': 'APE0639',\n",
       " '14600859': 'APE0640',\n",
       " '14600860': 'APE0641',\n",
       " '14600861': 'APE0643',\n",
       " '14600862': 'APE0644',\n",
       " '14600863': 'APE0647',\n",
       " '14600864': 'APE0648',\n",
       " '14600866': 'APE0650',\n",
       " '14600867': 'APES028',\n",
       " '14600868': 'APE0654',\n",
       " '14600869': 'APE0655',\n",
       " '14600870': 'APE0656',\n",
       " '14602193': 'APE0659a',\n",
       " '14600871': 'APE0660',\n",
       " '14600872': 'APE0661',\n",
       " '14600873': 'APE0662',\n",
       " '14600874': 'APE0663',\n",
       " '14600875': 'APE0665',\n",
       " '14600876': 'APES029',\n",
       " '14600877': 'APE0667',\n",
       " '14600878': 'APE0669',\n",
       " '14600879': 'APE0671',\n",
       " '14600880': 'APE0672',\n",
       " '14600881': 'APE0673',\n",
       " '14600882': 'APE0674',\n",
       " '14600883': 'APE0675',\n",
       " '14600884': 'APE0676',\n",
       " '14600885': 'APE0677',\n",
       " '14600887': 'APE0680',\n",
       " '14600888': 'APES031',\n",
       " '14600889': 'APE0681',\n",
       " '14600890': 'APE0682',\n",
       " '14600891': 'APES032',\n",
       " '14600892': 'APE0683',\n",
       " '14600893': 'APE0685',\n",
       " '14600894': 'APE0686',\n",
       " '14600895': 'APE0688',\n",
       " '14600896': 'APE0689',\n",
       " '14600897': 'APE0691',\n",
       " '14600898': 'APE0692',\n",
       " '14600899': 'APE0693',\n",
       " '14600900': 'APE0694',\n",
       " '14600901': 'APE0695',\n",
       " '14600902': 'APE0697',\n",
       " '14600903': 'APE0698',\n",
       " '14600904': 'APE0700',\n",
       " '14600905': 'APE0701',\n",
       " '14600906': 'APE0702',\n",
       " '14600907': 'APE0703',\n",
       " '14600908': 'APE0706',\n",
       " '14600909': 'APE0708',\n",
       " '14600910': 'APE0710',\n",
       " '14600911': 'APE0712',\n",
       " '14600912': 'APE0715',\n",
       " '14600913': 'APE0716',\n",
       " '14600914': 'APE0718',\n",
       " '14600915': 'APE0719',\n",
       " '14600916': 'APE0720',\n",
       " '14600917': 'APE0721',\n",
       " '14600918': 'APE0722',\n",
       " '14600919': 'APE0723',\n",
       " '14600920': 'APE0724',\n",
       " '14600921': 'APE0725',\n",
       " '14600922': 'APE0726',\n",
       " '14600923': 'APE0727',\n",
       " '14600924': 'APE0728',\n",
       " '14600925': 'APE0729',\n",
       " '14600926': 'APE0730',\n",
       " '14600927': 'APES033',\n",
       " '14600928': 'APE0731',\n",
       " '14600929': 'APE0732',\n",
       " '14600930': 'APE0733',\n",
       " '14600931': 'APE0734',\n",
       " '14600932': 'APE0735',\n",
       " '14600933': 'APE0736',\n",
       " '14600934': 'APE0737',\n",
       " '14600935': 'APE0739',\n",
       " '14600936': 'APE0741',\n",
       " '14600937': 'APE0743',\n",
       " '14600938': 'APE0744',\n",
       " '14600939': 'APE0745',\n",
       " '14600940': 'APE0748',\n",
       " '14602194': 'APE0749a',\n",
       " '14600941': 'APE0751',\n",
       " '14600942': 'APE0754',\n",
       " '14600944': 'APE0756',\n",
       " '14600945': 'APE0759',\n",
       " '14600946': 'APE0760',\n",
       " '14600947': 'APE0761',\n",
       " '14600948': 'APE0762',\n",
       " '14600949': 'APES034',\n",
       " '14600951': 'APE0766',\n",
       " '14600952': 'APE0767',\n",
       " '14600953': 'APE0768',\n",
       " '14600955': 'APE0771',\n",
       " '14600956': 'APE0772',\n",
       " '14600957': 'APE0773',\n",
       " '14600958': 'APE0774',\n",
       " '14600959': 'APE0775',\n",
       " '14600960': 'APE0776',\n",
       " '14600961': 'APE0777',\n",
       " '14600962': 'APE0778',\n",
       " '14600963': 'APE0781',\n",
       " '14600964': 'APE0782',\n",
       " '14600965': 'APE0784',\n",
       " '14600966': 'APE0785',\n",
       " '14600967': 'APE0786',\n",
       " '14600968': 'APE0787',\n",
       " '14600969': 'APE0788',\n",
       " '14600970': 'APE0789',\n",
       " '14600971': 'APE0790',\n",
       " '14600972': 'APE0791',\n",
       " '14600973': 'APE0792',\n",
       " '14600974': 'APE0793',\n",
       " '14600976': 'APE0796',\n",
       " '14600977': 'APE0797',\n",
       " '14600978': 'APE0798',\n",
       " '14600979': 'APE0800',\n",
       " '14600980': 'APE0801',\n",
       " '14600981': 'APE0803',\n",
       " '14600982': 'APE0805',\n",
       " '14600983': 'APE0807',\n",
       " '14600984': 'APE0809',\n",
       " '14600985': 'APE0812',\n",
       " '14600986': 'APE0813',\n",
       " '14600988': 'APE0815',\n",
       " '14600989': 'APE0816',\n",
       " '14600990': 'APE0818',\n",
       " '14600991': 'APE0819',\n",
       " '14600992': 'APE0820',\n",
       " '14600993': 'APE0821',\n",
       " '14600994': 'APE0822',\n",
       " '14600995': 'APE0824',\n",
       " '14600996': 'APE0825',\n",
       " '14600997': 'APE0826',\n",
       " '14600998': 'APES036',\n",
       " '14600999': 'APE0827',\n",
       " '14601000': 'APE0828',\n",
       " '14601001': 'APE0830',\n",
       " '14601002': 'APE0831',\n",
       " '14601003': 'APE0832',\n",
       " '14601004': 'APE0833',\n",
       " '14601005': 'APE0836',\n",
       " '14601006': 'APE0837',\n",
       " '14601007': 'APE0838',\n",
       " '14601008': 'APE0839',\n",
       " '14601009': 'APE0840',\n",
       " '14601010': 'APE0841',\n",
       " '14601011': 'APE0842',\n",
       " '14601012': 'APE0843',\n",
       " '14601013': 'APE0845',\n",
       " '14601014': 'APE0847',\n",
       " '14601015': 'APE0848',\n",
       " '14601016': 'APE0849',\n",
       " '14601017': 'APE0850',\n",
       " '14601018': 'APE0851',\n",
       " '14601019': 'APE0852',\n",
       " '14601020': 'APE0853',\n",
       " '14601021': 'APE0854',\n",
       " '14601022': 'APE0855',\n",
       " '14601023': 'APE0856',\n",
       " '14601024': 'APE0858',\n",
       " '14601025': 'APE0860',\n",
       " '14601026': 'APE0861',\n",
       " '14601027': 'APE0862',\n",
       " '14601028': 'APE0864',\n",
       " '14601029': 'APE0865',\n",
       " '14601030': 'APE0866',\n",
       " '14601031': 'APE0867',\n",
       " '14601032': 'APE0868',\n",
       " '14601033': 'APE0869',\n",
       " '14601034': 'APES037',\n",
       " '14601035': 'APE0870',\n",
       " '14601036': 'APES038',\n",
       " '14601037': 'APE0872',\n",
       " '14601038': 'APE0874',\n",
       " '14601039': 'APE0875',\n",
       " '14601040': 'APE0878',\n",
       " '14601041': 'APE0879',\n",
       " '14601042': 'APE0880',\n",
       " '14601043': 'APES039',\n",
       " '14601044': 'APE0881',\n",
       " '14601045': 'APE0882',\n",
       " '14601046': 'APE0883',\n",
       " '14601047': 'APE0884',\n",
       " '14601048': 'APE0885',\n",
       " '14601049': 'APE0886',\n",
       " '14601050': 'APE0887',\n",
       " '14601051': 'APE0888',\n",
       " '14601052': 'APE0889',\n",
       " '14601053': 'APE0890',\n",
       " '14601054': 'APES040',\n",
       " '14601055': 'APE0891',\n",
       " '14601056': 'APE0892',\n",
       " '14601057': 'APE0894',\n",
       " '14601058': 'APE0895',\n",
       " '14601059': 'APE0896',\n",
       " '14601060': 'APE0898',\n",
       " '14601061': 'APE0900',\n",
       " '14601062': 'APE0902',\n",
       " '14601063': 'APE0903',\n",
       " '14601064': 'APE0906',\n",
       " '14601065': 'APE0907',\n",
       " '14601066': 'APE0912',\n",
       " '14601067': 'APE0914',\n",
       " '14601068': 'APE0915',\n",
       " '14601069': 'APE0916',\n",
       " '14601070': 'APE0917',\n",
       " '14601071': 'APE0919',\n",
       " '14601072': 'APE0921',\n",
       " '14601073': 'APE0922',\n",
       " '14601074': 'APE0924',\n",
       " '14601076': 'APE0927',\n",
       " '14601077': 'APE0928',\n",
       " '14601078': 'APE0929',\n",
       " '14601079': 'APE0930',\n",
       " '14601080': 'APE0931',\n",
       " '14601081': 'APE0932',\n",
       " '14601082': 'APE0933',\n",
       " '14601083': 'APE0934',\n",
       " '14601084': 'APE0935',\n",
       " '14601085': 'APE0938',\n",
       " '14601086': 'APE0939',\n",
       " '14601087': 'APE0940',\n",
       " '14601088': 'APE0941',\n",
       " '14601089': 'APE0943',\n",
       " '14601090': 'APE0944',\n",
       " '14601091': 'APE0945',\n",
       " '14601092': 'APE0946',\n",
       " '14601093': 'APE0947',\n",
       " '14601094': 'APE0948',\n",
       " '14601095': 'APE0950',\n",
       " '14601096': 'APE0951',\n",
       " '14601097': 'APE0952',\n",
       " '14601098': 'APE0954',\n",
       " '14601099': 'APE0955',\n",
       " '14601101': 'APE0957',\n",
       " '14602195': 'APE0958a',\n",
       " '14601102': 'APE0959',\n",
       " '14601103': 'APE0960',\n",
       " '14601104': 'APE0964',\n",
       " '14601105': 'APE0965',\n",
       " '14601106': 'APE0966',\n",
       " '14601107': 'APE0967',\n",
       " '14601108': 'APE0969',\n",
       " '14601109': 'APE0971',\n",
       " '14601110': 'APE0972',\n",
       " '14601111': 'APE0973',\n",
       " '14601112': 'APE0974',\n",
       " '14601113': 'APE0975',\n",
       " '14601114': 'APE0976',\n",
       " '14601115': 'APE0978',\n",
       " '14602196': 'APE0978a',\n",
       " '14601116': 'APE0979',\n",
       " '14601117': 'APE0980',\n",
       " '14601118': 'APE0981',\n",
       " '14601119': 'APE0983',\n",
       " '14601120': 'APE0985',\n",
       " '14601121': 'APE0986',\n",
       " '14601122': 'APE0988',\n",
       " '14601123': 'APE0989',\n",
       " '14601124': 'APE0990',\n",
       " '14601125': 'APE0993',\n",
       " '14601126': 'APE0994',\n",
       " '14601127': 'APE0996',\n",
       " '14601128': 'APE1000',\n",
       " '14601129': 'APE1001',\n",
       " '14601130': 'APE1002',\n",
       " '14601131': 'APE1003',\n",
       " '14601132': 'APE1006',\n",
       " '14601133': 'APE1007',\n",
       " '14601134': 'APE1008',\n",
       " '14601135': 'APES041',\n",
       " '14601136': 'APE1009',\n",
       " '14601138': 'APE1011',\n",
       " '14601139': 'APE1013',\n",
       " '14601140': 'APE1014',\n",
       " '14601141': 'APE1015',\n",
       " '14601143': 'APE1023',\n",
       " '14601145': 'APE1026',\n",
       " '14601146': 'APE1029',\n",
       " '14601147': 'APE1031',\n",
       " '14601148': 'APE1035',\n",
       " '14601149': 'APE1038',\n",
       " '14601150': 'APE1040',\n",
       " '14601151': 'APE1041',\n",
       " '14601152': 'APE1042',\n",
       " '14602197': 'APE1043m',\n",
       " '14601153': 'APE1044',\n",
       " '14601154': 'APE1045',\n",
       " '14601156': 'APE1047',\n",
       " '14601157': 'APE1049',\n",
       " '14601158': 'APE1050',\n",
       " '14601160': 'APE1053',\n",
       " '14601161': 'APE1055',\n",
       " '14601162': 'APE1056',\n",
       " '14601163': 'APE1059',\n",
       " '14601164': 'APE1061',\n",
       " '14601165': 'APE1063',\n",
       " '14601166': 'APE1064',\n",
       " '14601167': 'APE1065',\n",
       " '14601168': 'APE1072',\n",
       " '14601169': 'APE1073',\n",
       " '14601171': 'APE1078',\n",
       " '14601172': 'APE1080',\n",
       " '14601173': 'APE1081',\n",
       " '14601174': 'APE1084',\n",
       " '14602198': 'APE1084a',\n",
       " '14601175': 'APE1085',\n",
       " '14601176': 'APE1087',\n",
       " '14602199': 'APE1087a',\n",
       " '14602200': 'APE1088a',\n",
       " '14601178': 'APE1090',\n",
       " '14601179': 'APE1091',\n",
       " '14601180': 'APE1092',\n",
       " '14601181': 'APE1094',\n",
       " '14601183': 'APE1098',\n",
       " '14601184': 'APE1102',\n",
       " '14601185': 'APE1104',\n",
       " '14601186': 'APE1106',\n",
       " '14601187': 'APE1109',\n",
       " '14601188': 'APE1113',\n",
       " '14601189': 'APE1114',\n",
       " '14601190': 'APE1117',\n",
       " '14601191': 'APE1119',\n",
       " '14601193': 'APE1123',\n",
       " '14601194': 'APE1124',\n",
       " '14601195': 'APE1125',\n",
       " '14601196': 'APE1126',\n",
       " '14601197': 'APE1127',\n",
       " '14601199': 'APE1129',\n",
       " '14601200': 'APE1132',\n",
       " '14602201': 'APE1132a',\n",
       " '14601201': 'APE1135',\n",
       " '14601202': 'APE1136',\n",
       " '14601203': 'APE1138',\n",
       " '14601204': 'APE1139',\n",
       " '14601205': 'APE1140',\n",
       " '14601206': 'APE1143',\n",
       " '14601207': 'APE1144',\n",
       " '14601208': 'APE1147',\n",
       " '14601209': 'APE1148',\n",
       " '14601210': 'APE1149',\n",
       " '14601211': 'APE1150',\n",
       " '14601212': 'APE1151',\n",
       " '14601213': 'APE1152',\n",
       " '14601215': 'APE1154',\n",
       " '14601216': 'APE1156',\n",
       " '14601217': 'APE1158',\n",
       " '14601218': 'APE1161',\n",
       " '14601219': 'APE1164',\n",
       " '14601220': 'APE1167',\n",
       " '14601221': 'APE1169',\n",
       " '14601222': 'APE1171',\n",
       " '14601223': 'APE1173',\n",
       " '14601224': 'APE1174',\n",
       " '14601225': 'APE1176',\n",
       " '14601226': 'APE1177',\n",
       " '14601227': 'APES042',\n",
       " '14601228': 'APE1178',\n",
       " '14601229': 'APE1179',\n",
       " '14601230': 'APE1180',\n",
       " '14601231': 'APE1181',\n",
       " '14601232': 'APE1182',\n",
       " '14601233': 'APE1183',\n",
       " '14601234': 'APE1184',\n",
       " '14601235': 'APE1186',\n",
       " '14601236': 'APE1187',\n",
       " '14601237': 'APE1188',\n",
       " '14602202': 'APE1189m',\n",
       " '14601238': 'APE1190',\n",
       " '14601239': 'APE1191',\n",
       " '14601240': 'APE1192',\n",
       " '14601241': 'APE1193',\n",
       " '14601242': 'APE1194',\n",
       " '14601243': 'APE1195',\n",
       " '14601244': 'APE1196',\n",
       " '14601245': 'APE1197',\n",
       " '14601246': 'APE1199',\n",
       " '14601247': 'APE1202',\n",
       " '14601249': 'APE1205',\n",
       " '14601250': 'APE1208',\n",
       " '14601251': 'APE1209',\n",
       " '14602203': 'APE1209a',\n",
       " '14601252': 'APES043',\n",
       " '14601253': 'APES044',\n",
       " '14601254': 'APES045',\n",
       " '14601255': 'APE1210',\n",
       " '14601256': 'APE1211',\n",
       " '14601257': 'APE1212',\n",
       " '14601258': 'APE1213',\n",
       " '14601260': 'APE1215',\n",
       " '14601261': 'APE1216',\n",
       " '14601262': 'APE1217',\n",
       " '14601263': 'APE1219',\n",
       " '14601264': 'APE1220',\n",
       " '14601265': 'APE1221',\n",
       " '14601266': 'APE1222',\n",
       " '14601267': 'APE1223',\n",
       " '14601268': 'APE1226',\n",
       " '14601269': 'APE1228',\n",
       " '14601270': 'APE1229',\n",
       " '14601271': 'APE1230',\n",
       " '14601272': 'APE1231',\n",
       " '14601273': 'APE1232',\n",
       " '14601275': 'APE1234',\n",
       " '14601276': 'APE1235',\n",
       " '14601277': 'APE1236',\n",
       " '14601278': 'APE1237',\n",
       " '14601279': 'APE1239',\n",
       " '14601280': 'APE1240',\n",
       " '14601281': 'APE1241',\n",
       " '14601282': 'APES046',\n",
       " '14601283': 'APE1242',\n",
       " '14601284': 'APE1243',\n",
       " '14601285': 'APE1244',\n",
       " '14601286': 'APE1245',\n",
       " '14601287': 'APE1248',\n",
       " '14601288': 'APE1250',\n",
       " '14601289': 'APE1251',\n",
       " '14601290': 'APE1252',\n",
       " '14601291': 'APE1253',\n",
       " '14601292': 'APE1255',\n",
       " '14601293': 'APE1257',\n",
       " '14601294': 'APE1258',\n",
       " '14601295': 'APE1260',\n",
       " '14601296': 'APE1262',\n",
       " '14601298': 'APE1265',\n",
       " '14601299': 'APE1267',\n",
       " '14601300': 'APE1270',\n",
       " '14601301': 'APE1272',\n",
       " '14601302': 'APE1273',\n",
       " '14601303': 'APE1275',\n",
       " '14601304': 'APES047',\n",
       " '14601305': 'APE1276',\n",
       " '14601306': 'APE1277',\n",
       " '14601307': 'APE1278',\n",
       " '14601308': 'APE1279',\n",
       " '14601309': 'APE1280',\n",
       " '14601310': 'APE1281',\n",
       " '14601311': 'APE1284',\n",
       " '14601312': 'APE1285',\n",
       " '14601313': 'APE1286',\n",
       " '14601314': 'APE1288',\n",
       " '14601315': 'APE1294',\n",
       " '14601316': 'APE1297',\n",
       " '14601317': 'APE1298',\n",
       " '14601318': 'APE1299',\n",
       " '14601319': 'APE1300',\n",
       " '14601320': 'APE1301',\n",
       " '14601322': 'APE1303',\n",
       " '14601323': 'APE1304',\n",
       " ...}"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_id_dic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'k is less than '"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mode = \"1 3\"\n",
    "if \"1\" in mode:\n",
    "    for i in user_selected_number:\n",
    "        for k in  user_selected_number:\n",
    "            if k<i:\n",
    "                display(\"k is less than \")\n",
    "                pass\n",
    "            elif i == k :\n",
    "                read_equal_bbh(\n",
    "                score_file + selected_species_dic[i] +)\n",
    "                \n",
    "            \n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "equal_BBH = []\n",
    "second_equal_BBH_data = []\n",
    "with open(\"./score_file/AAE_AAE_oneway_threshold_best_hit_Score5\") as equal_RBH:\n",
    "    for j in equal_RBH:\n",
    "        split_data = j.split()\n",
    "        split_data[2] = int(split_data[2])\n",
    "        equal_BBH.append(split_data)\n",
    "try:\n",
    "    with open(\"./score_file/AAE_AAE_second_oneway_threshold_best_hit_Score5\") as second_equal_RBH:\n",
    "        for j in second_equal_RBH:\n",
    "            split_data = j.split()\n",
    "            # \"[]\"\n",
    "            split_data[2] = int(split_data[2])\n",
    "            second_equal_BBH_data.append(split_data)\n",
    "except IOError as err:\n",
    "    print(err)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['AAE_gi|15606877|ref|NP_214257.1|', 'AAE_gi|15606128|ref|NP_213505.1|', 898],\n",
       " ['AAE_gi|15606581|ref|NP_213961.1|',\n",
       "  'AAE_gi|15606744|ref|NP_214124.1|',\n",
       "  2170],\n",
       " ['AAE_gi|15606128|ref|NP_213505.1|', 'AAE_gi|15606877|ref|NP_214257.1|', 898],\n",
       " ['AAE_gi|15606744|ref|NP_214124.1|',\n",
       "  'AAE_gi|15606581|ref|NP_213961.1|',\n",
       "  2170],\n",
       " ['AAE_gi|15606877|ref|NP_214257.1|', 'AAE_gi|15606128|ref|NP_213505.1|', 898],\n",
       " ['AAE_gi|15606581|ref|NP_213961.1|',\n",
       "  'AAE_gi|15606744|ref|NP_214124.1|',\n",
       "  2170],\n",
       " ['AAE_gi|15606128|ref|NP_213505.1|', 'AAE_gi|15606877|ref|NP_214257.1|', 898],\n",
       " ['AAE_gi|15606744|ref|NP_214124.1|',\n",
       "  'AAE_gi|15606581|ref|NP_213961.1|',\n",
       "  2170],\n",
       " ['AAE_gi|15606877|ref|NP_214257.1|', 'AAE_gi|15606128|ref|NP_213505.1|', 609],\n",
       " ['AAE_gi|15606581|ref|NP_213961.1|',\n",
       "  'AAE_gi|15606744|ref|NP_214124.1|',\n",
       "  1559],\n",
       " ['AAE_gi|15606128|ref|NP_213505.1|', 'AAE_gi|15606877|ref|NP_214257.1|', 609],\n",
       " ['AAE_gi|15606744|ref|NP_214124.1|',\n",
       "  'AAE_gi|15606581|ref|NP_213961.1|',\n",
       "  1559]]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mode 2 One_way_threshold_\n",
    "for i in user_selected_number:\n",
    "    query_v = query_seq_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P\n",
      "y\n",
      "t\n",
      "o\n",
      "n\n"
     ]
    }
   ],
   "source": [
    "for letter in \"Python\":\n",
    "    if letter == \"h\":\n",
    "        \n",
    "        continue\n",
    "    print(letter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "nbTranslate": {
   "displayLangs": [
    "ko"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "en",
   "targetLang": "ko",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
